{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import yaml\n",
    "\n",
    "#Read options file\n",
    "with open(\"optionsRscript.yaml\", 'r') as stream:\n",
    "    data = yaml.safe_load(stream)\n",
    "\n",
    "homeDir = os.getcwd()+\"/\"\n",
    "\n",
    "homeDir = \"/storage/gluster/vol1/SHARED/HOMEFOLDERS/adefalco/\"\n",
    "fileDir = homeDir+\"Analysis/\"\n",
    "\n",
    "fpfilterDir = fileDir + \"fpfilter/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: \n",
      "   *****       ***   vcfR   ***       *****\n",
      "   This is vcfR 1.12.0 \n",
      "     browseVignettes('vcfR') # Documentation\n",
      "     citation('vcfR') # Citation\n",
      "   *****       *****      *****       *****\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: foreach\n",
      "\n",
      "R[write to console]: Loading required package: iterators\n",
      "\n",
      "R[write to console]: Loading required package: parallel\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 306237\n"
     ]
    }
   ],
   "source": [
    "%%R -i fileDir -i workDir -i vcfDir -i vcfFile\n",
    "#devtools::install_github(repo=\"knausb/vcfR\")\n",
    "#install.packages('doParallel')\n",
    "library(vcfR)\n",
    "library(doParallel)\n",
    "#setwd(\"/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/\")\n",
    "dir.create(fileDir)\n",
    "setwd(workDir)\n",
    "\n",
    "# VCF-------------------\n",
    "filenames <- dir(vcfDir,vcfFile)\n",
    "#vcfDir <- \"ensemble/vcf/\" MY\n",
    "#filenames <- dir(vcfDir,\"ensemble.snpEff.vcf.gz\")\n",
    "\n",
    "#filenames <- filenames[!grepl(\".tbi\",filenames)]\n",
    "\n",
    "vcfRead <- function(filename,vcfDir){\n",
    "  tumorID <- sub(paste(\"-\",vcfFile,sep=\"\"),\"\",filename) #\"-ensemble.snpEff.vcf.gz\"\n",
    "  tmp <- vcfR::read.vcfR(paste0(vcfDir,filename), verbose=FALSE)\n",
    "  tmpVcf <- tmp@fix\n",
    "  tmpGt <- tmp@gt\n",
    "  if(length(colnames(tmpGt))==3){\n",
    "    if(!grepl(\"_N\\\\b\",colnames(tmpGt)[3])){ #CHECK THE NAME OF TUMOR AND NORMAL SAMPLES IN VCF!\n",
    "      tmpGt <- tmpGt[,c(1,3,2)]\n",
    "    }\n",
    "  }\n",
    "  info <- tmpVcf[,\"INFO\"]\n",
    "  func <- sapply(info,function(x){\n",
    "    temp <- unlist(strsplit(x,\",\"))\n",
    "    paste0(unique(na.omit(unlist(sapply(temp,function(x) unlist(strsplit(x,\"|\",fixed=TRUE))[2])))),collapse=\",\")\n",
    "  })\n",
    "  func <- unname(func)\n",
    "  callers <- sapply(info,function(x){\n",
    "    temp <- unlist(strsplit(x,\";\"))\n",
    "    sub(\"CALLERS=\",\"\",temp[grepl(\"CALLERS=\",temp)])\n",
    "  })\n",
    "  callers <- unname(callers)\n",
    "  id <- paste0(tumorID,\"-\",tmpVcf[,1],\":\",tmpVcf[,2])\n",
    "  if(length(colnames(tmpGt))==3){\n",
    "    tmp <- data.frame(tmpVcf[,c(1,2,4,5,7)],tmpGt,tumorID,info,func,callers,id,stringsAsFactors=FALSE)\n",
    "  } else {\n",
    "    tmp <- data.frame(tmpVcf[,c(1,2,4,5,7)],tmpGt,normal=NA,tumorID,info,func,callers,id,stringsAsFactors=FALSE)\n",
    "  }\n",
    "  colnames(tmp) <- c(\"CHROM\",\"POS\",\"REF\",\"ALT\",\"FILTER\",\"FORMAT\",\"Tumor\",\"Normal\",\"Tumor_ID\",\"snpEff\",\"snpEff_func\",\"caller\",\"id\")\n",
    "  rownames(tmp) <- NULL\n",
    "  return(tmp)\n",
    "}\n",
    "\n",
    "vcf <- lapply(filenames,function(x) vcfRead(x,vcfDir))\n",
    "vcf <- do.call(rbind,vcf)\n",
    "nrow(vcf) #306237\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CORRECT VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "# CORRECT FOR NA ALT\n",
    "table(is.na(vcf$ALT))\n",
    "# vcf <- vcf[!is.na(vcf$ALT),]\n",
    "\n",
    "# CORRECT FOR MULTIPLE ALT\n",
    "table(grepl(\",\",vcf$ALT))\n",
    "tmp <- sapply(vcf$ALT,function(x) unlist(strsplit(x,\",\",fixed=TRUE))[1])\n",
    "vcf$ALT <- tmp\n",
    "\n",
    "# # CORRECT FOR NA LINES\n",
    "#sum(is.na(vcf))\n",
    "#vcf <- vcf[!is.na(vcf)]\n",
    "#sum(is.na(vcf))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SELECT CODING/SPLICING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i fileDir\n",
    "# SELECT CODING/SPLICING\n",
    "tmp <- unique(vcf$snpEff_func)\n",
    "tmp <- unique(unlist(sapply(tmp,function(x) unlist(strsplit(x,\",\")))))\n",
    "tmp <- unique(unlist(sapply(tmp,function(x) unlist(strsplit(x,\"&\")))))\n",
    "tmp <- tmp[!grepl(\"ENS\",tmp)]\n",
    "tmp <- tmp[!grepl(\"=\",tmp)]\n",
    "sort(tmp)\n",
    "sel <- c(\"5_prime_UTR_premature_start_codon_gain_variant\",\"5_prime_UTR_truncation\", #NEW\n",
    "         \"3_prime_UTR_variant\", \"5_prime_UTR_variant\", # Added\n",
    "         \"bidirectional_gene_fusion\",\"conservative_inframe_deletion\",\"conservative_inframe_insertion\",\n",
    "         \"disruptive_inframe_deletion\",\"disruptive_inframe_insertion\",\n",
    "         \"exon_loss_variant\",\"initiator_codon_variant\",\"gene_fusion\",\n",
    "         \"frameshift_variant\",\"missense_variant\",\"splice_acceptor_variant\",\n",
    "         \"splice_donor_variant\",\"splice_region_variant\",\"start_lost\",\"stop_gained\",\n",
    "         \"stop_lost\",\"stop_retained_variant\")\n",
    "setdiff(tmp,sel)\n",
    "setdiff(sel,tmp)\n",
    "sel <- sapply(sel,function(x) grep(x,vcf$snpEff_func))\n",
    "sel <- unique(unlist(sel))\n",
    "sel <- sort(sel)\n",
    "sum(is.na(sel))\n",
    "nrow(vcf) #227168\n",
    "rownames(vcf) <- NULL\n",
    "length(sel) #36938\n",
    "vcf <- vcf[sel,]\n",
    "sort(table(vcf$Tumor_ID))\n",
    "\n",
    "callerN <- sapply(vcf$caller,function(x) length(unlist(strsplit(x,\",\",fixed=TRUE))))\n",
    "callerN <- unname(callerN)\n",
    "vcf <- data.frame(vcf,callerN,stringsAsFactors = FALSE)\n",
    "table(vcf$callerN)\n",
    "#   1     2     3     4     5     6 \n",
    "# 29690  3753   704   328   438  2025  \n",
    "rownames(vcf) <- NULL\n",
    "#----------------------------------------------\n",
    "\n",
    "save(vcf,file = paste(fileDir,\"vcf.rda\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COUNTS and ALLELE FREQUENCY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 27894\n"
     ]
    }
   ],
   "source": [
    "%%R -o counts\n",
    "# COUNTS and ALLELE FREQUENCY -----------------\n",
    "\n",
    "countsVcf <- function(vcf,i){\n",
    "  Tmp <- rep(NA,4)\n",
    "  names(Tmp) <- c(\"t_ref_count\",\"t_alt_count\",\"n_ref_count\",\"n_alt_count\")\n",
    "  vcfLine <- as.character(vcf[i,])\n",
    "  names(vcfLine) <- colnames(vcf)\n",
    "  caller <- unlist(strsplit(vcfLine[\"caller\"],\",\",fixed=TRUE))[1]\n",
    "  if(caller==\"varscan\"){\n",
    "    tmp <- unlist(strsplit(vcfLine[\"Tumor\"],\":\"))\n",
    "    names(tmp) <- unlist(strsplit(vcfLine[\"FORMAT\"],\":\"))\n",
    "    Tmp[\"t_alt_count\"] <- as.numeric(tmp[\"AD\"])\n",
    "    Tmp[\"t_ref_count\"] <- as.numeric(tmp[\"RD\"])\n",
    "    tmp <- unlist(strsplit(vcfLine[\"Normal\"],\":\"))\n",
    "    names(tmp) <- unlist(strsplit(vcfLine[\"FORMAT\"],\":\"))\n",
    "    Tmp[\"n_alt_count\"] <- as.numeric(tmp[\"AD\"])\n",
    "    Tmp[\"n_ref_count\"] <- as.numeric(tmp[\"RD\"])\n",
    "  } else {\n",
    "    tmp <- unlist(strsplit(vcfLine[\"Tumor\"],\":\"))\n",
    "    names(tmp) <- unlist(strsplit(vcfLine[\"FORMAT\"],\":\"))\n",
    "    if(tmp[\"AD\"]!=\".\") Tmp[c(\"t_ref_count\",\"t_alt_count\")] <- as.numeric(unlist(strsplit(tmp[\"AD\"],\",\"))[1:2])\n",
    "    if(!is.na(vcfLine[\"Normal\"])) {\n",
    "      tmp <- unlist(strsplit(vcfLine[\"Normal\"],\":\"))\n",
    "      names(tmp) <- unlist(strsplit(vcfLine[\"FORMAT\"],\":\"))[1:length(tmp)]\n",
    "      if(tmp[\"AD\"]!=\".\") Tmp[c(\"n_ref_count\",\"n_alt_count\")] <- as.numeric(unlist(strsplit(tmp[\"AD\"],\",\"))[1:2])\n",
    "    }\n",
    "  }\n",
    "  return(Tmp)\n",
    "}\n",
    "\n",
    "cl <- makePSOCKcluster(60,outfile=\"\")\n",
    "registerDoParallel(cl)\n",
    "tmp <- foreach(i = 1:nrow(vcf)) %dopar% {\n",
    "  tmp <- countsVcf(vcf,i)\n",
    "  return(tmp)\n",
    "}\n",
    "stopCluster(cl)\n",
    "\n",
    "tmp <- do.call(rbind,tmp)\n",
    "dim(tmp) #36938     4\n",
    "\n",
    "counts <- cbind(tmp[,1:2],NA,NA,tmp[,3:4],NA,NA)\n",
    "colnames(counts) <- c(\"t_ref_count\",\"t_alt_count\",\"t_depth\",\"t_vaf\",\"n_ref_count\",\"n_alt_count\",\"n_depth\",\"n_vaf\")\n",
    "counts[,\"t_depth\"] <- counts[,\"t_alt_count\"] + counts[,\"t_ref_count\"]\n",
    "counts[,\"t_vaf\"] <- round(counts[,\"t_alt_count\"]/counts[,\"t_depth\"],2)\n",
    "counts[,\"n_depth\"] <- counts[,\"n_alt_count\"] + counts[,\"n_ref_count\"]\n",
    "counts[,\"n_vaf\"] <- round(counts[,\"n_alt_count\"]/counts[,\"n_depth\"],2)\n",
    "rownames(counts) <- NULL\n",
    "\n",
    "#View(counts[sample(1:nrow(counts),100),])\n",
    "\n",
    "vcf <- data.frame(vcf, counts, stringsAsFactors = FALSE)\n",
    "\n",
    "table(cond <- is.na(vcf$n_depth)&!is.na(vcf$Normal))\n",
    "# FALSE  TRUE \n",
    "# 27894  9044 \n",
    "# table(cond <- is.na(vcf$n_depth))\n",
    "\n",
    "tmp <- vcf[cond,]\n",
    "\n",
    "# table(tmp$caller)\n",
    "# freebayes \n",
    "# 9044\n",
    "\n",
    "vcf <- vcf[!cond,]\n",
    "\n",
    "rownames(vcf) <- NULL\n",
    "nrow(vcf) #27894\n",
    "#rm(counts)\n",
    "\n",
    "#----------------------------------------------\n",
    "# save(vcf,file = \"Analysis/vcf.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.27</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>79.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>26.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.53</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.24</td>\n",
       "      <td>54.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>209.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27889</th>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27890</th>\n",
       "      <td>66.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>0.19</td>\n",
       "      <td>128.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>134.0</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27891</th>\n",
       "      <td>32.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>291.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>291.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27892</th>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27893</th>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.16</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27894 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0     1      2     3      4    5      6     7\n",
       "0        8.0   3.0   11.0  0.27   67.0  0.0   67.0  0.00\n",
       "1       79.0   3.0   82.0  0.04  179.0  0.0  179.0  0.00\n",
       "2       26.0  29.0   55.0  0.53  211.0  0.0  211.0  0.00\n",
       "3       16.0   5.0   21.0  0.24   54.0  3.0   57.0  0.05\n",
       "4      118.0   8.0  126.0  0.06  209.0  1.0  210.0  0.00\n",
       "...      ...   ...    ...   ...    ...  ...    ...   ...\n",
       "27889    0.0  17.0   17.0  1.00    8.0  0.0    8.0  0.00\n",
       "27890   66.0  15.0   81.0  0.19  128.0  6.0  134.0  0.04\n",
       "27891   32.0  46.0   78.0  0.59  291.0  0.0  291.0  0.00\n",
       "27892    0.0  43.0   43.0  1.00    5.0  0.0    5.0  0.00\n",
       "27893   16.0   3.0   19.0  0.16   11.0  0.0   11.0  0.00\n",
       "\n",
       "[27894 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "import pandas as pd\n",
    "pd.set_option('display.min_rows',10)\n",
    "pd.set_option('display.max_rows',10)\n",
    "countsp = pd.DataFrame(data=counts)#,    # values\n",
    "    #index=counts[1:,0],    # 1st column as index\n",
    "    #columns=counts[0,1:]) \n",
    "display(countsp)\n",
    "#print (\"Dataframe 2:\")\n",
    "#display(HTML(countsp.to_html()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FP FILTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"A1783\"\n",
      "[1] \"A9155\"\n",
      "[1] \"A9878\"\n",
      "[1] \"AT2550\"\n",
      "[1] \"AT2822\"\n",
      "[1] \"AT4415\"\n",
      "[1] \"AT4808\"\n",
      "[1] \"C0288\"\n",
      "[1] \"C0334\"\n",
      "[1] \"C1572\"\n",
      "[1] \"C8448\"\n",
      "[1] \"s1_3076_D\"\n",
      "[1] \"s1_3295_D\"\n",
      "[1] \"s1_8588_D\"\n",
      "[1] \"s2_6088_D\"\n",
      "[1] \"s2_6339_D\"\n",
      "[1] \"s2_8811_D\"\n",
      "[1] \"s3_5832_D\"\n",
      "[1] \"s3_7299_D\"\n",
      "[1] \"s3_7979_D\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i fpfilterDir\n",
    "# FP FILTER\n",
    "#system(\"mkdir Analysis/fpfilter\")\n",
    "#setwd(\"Analysis/fpfilter/\")\n",
    "dir.create(fpfilterDir)\n",
    "setwd(fpfilterDir)\n",
    "\n",
    "vcf <- data.frame(vcf,fpFid=\"\",stringsAsFactors=FALSE)\n",
    "patients <- unique(vcf$Tumor_ID)\n",
    "for(i in 1:length(patients)){\n",
    "  print(patients[i])\n",
    "  name <- patients[i]\n",
    "  var <- vcf[vcf$Tumor_ID==name,1:4]\n",
    "  var[nchar(var$ALT)>1,\"ALT\"] <- paste0(\"+\",substring(var[nchar(var$ALT)>1,\"ALT\"],2))\n",
    "  var[nchar(var$REF)>1,\"POS\"] <- as.numeric(var[nchar(var$REF)>1,\"POS\"])+1\n",
    "  var[nchar(var$REF)>1,\"ALT\"] <- paste0(\"-\",substring(var[nchar(var$REF)>1,\"REF\"],2))\n",
    "  var[nchar(var$REF)>1,\"REF\"] <- substring(var[nchar(var$REF)>1,\"REF\"],2,2)\n",
    "  vcf[vcf$Tumor_ID==name,\"fpFid\"] <- paste0(name,\"-\",var[,1],\":\",var[,2],\"-\",var[,4])\n",
    "  write.table(var,file=paste0(name,\".var\"),sep=\"\\t\",quote=FALSE,row.names=FALSE,col.names=FALSE)\n",
    "  # n <- ceiling(nrow(var)/60)\n",
    "  # system(paste0(\"split -l \",n,\" --additional-suffix .var \",paste0(name,\".var\")))\n",
    "  loc <- paste(var$CHROM,var$POS,var$POS)\n",
    "  write.table(loc,file=paste0(name,\".loc\"),quote=FALSE,row.names=FALSE,col.names=FALSE)\n",
    "  # system(paste0(\"split -l \",n,\" --additional-suffix .loc \",paste0(name,\".loc\")))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(fpfilterDir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "myPythonVariable's value is /storage/gluster/vol1/bcbio/genomes/Hsapiens/hg19/bwa/hg19.fa and /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/\n",
      "A1783.loc\n",
      "A1783\n",
      "A1783 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/A1783_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/A1783_T_recal.bam\n",
      "\\n\n",
      "A9155.loc\n",
      "A9155\n",
      "A9155 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/A9155_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/A9155_T_recal.bam\n",
      "\\n\n",
      "A9878.loc\n",
      "A9878\n",
      "A9878 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/A9878_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/A9878_T_recal.bam\n",
      "\\n\n",
      "AT2550.loc\n",
      "AT2550\n",
      "AT2550 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/AT2550_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/AT2550_T_recal.bam\n",
      "\\n\n",
      "AT2822.loc\n",
      "AT2822\n",
      "AT2822 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/AT2822_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/AT2822_T_recal.bam\n",
      "\\n\n",
      "AT4415.loc\n",
      "AT4415\n",
      "AT4415 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/AT4415_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/AT4415_T_recal.bam\n",
      "\\n\n",
      "AT4808.loc\n",
      "AT4808\n",
      "AT4808 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/AT4808_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/AT4808_T_recal.bam\n",
      "\\n\n",
      "C0288.loc\n",
      "C0288\n",
      "C0288 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/C0288_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/C0288_T_recal.bam\n",
      "\\n\n",
      "C0334.loc\n",
      "C0334\n",
      "C0334 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/C0334_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/C0334_T_recal.bam\n",
      "\\n\n",
      "C1572.loc\n",
      "C1572\n",
      "C1572 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/C1572_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/C1572_T_recal.bam\n",
      "\\n\n",
      "C8448.loc\n",
      "C8448\n",
      "C8448 bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_Flank/ABT414_Flank/C8448_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/C8448_T_recal.bam\n",
      "\\n\n",
      "s1_3076_D.loc\n",
      "s1_3076_D\n",
      "s1_3076_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s1_3076_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s1_3076_D_T_recal.bam\n",
      "\\n\n",
      "s1_3295_D.loc\n",
      "s1_3295_D\n",
      "s1_3295_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s1_3295_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s1_3295_D_T_recal.bam\n",
      "\\n\n",
      "s1_8588_D.loc\n",
      "s1_8588_D\n",
      "s1_8588_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s1_8588_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s1_8588_D_T_recal.bam\n",
      "\\n\n",
      "s2_6088_D.loc\n",
      "s2_6088_D\n",
      "s2_6088_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s2_6088_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s2_6088_D_T_recal.bam\n",
      "\\n\n",
      "s2_6339_D.loc\n",
      "s2_6339_D\n",
      "s2_6339_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s2_6339_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s2_6339_D_T_recal.bam\n",
      "\\n\n",
      "s2_8811_D.loc\n",
      "s2_8811_D\n",
      "s2_8811_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s2_8811_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s2_8811_D_T_recal.bam\n",
      "\\n\n",
      "s3_5832_D.loc\n",
      "s3_5832_D\n",
      "s3_5832_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s3_5832_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s3_5832_D_T_recal.bam\n",
      "\\n\n",
      "s3_7299_D.loc\n",
      "s3_7299_D\n",
      "s3_7299_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s3_7299_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s3_7299_D_T_recal.bam\n",
      "\\n\n",
      "s3_7979_D.loc\n",
      "s3_7979_D\n",
      "s3_7979_D bam-readcount\n",
      "/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/ABT414_IC/ABT414_IC/s3_7979_D_T_recal.bam /storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/CNV/tumorbam/s3_7979_D_T_recal.bam\n",
      "\\n\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "Minimum mapping quality is set to 0\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1254:23402:13244: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1342:11595:20541: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:2169:31204:21809: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1110:25807:9643: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1453:20573:5149: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1465:12662:35008: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1412:30870:29371: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1119:8820:27696: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1375:8585:36025: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:1274:26096:16063: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:2253:4698:31610: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00261:111:HFJ5KDSXX:2:1217:1841:13698: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:1241:27606:5274: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:1134:9878:32471: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:2337:14886:17879: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:1637:19813:24064: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:2508:7545:22576: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:2660:19226:11804: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:1446:15854:20181: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n",
      "WARNING: In read A00197:72:HCHTWDSXX:2:1157:30933:31673: Couldn't find single-end mapping quality. Check to see if the SM tag is in BAM.\n",
      "The previous warning has been emitted 1 times and will be disabled.\n"
     ]
    }
   ],
   "source": [
    "%%script env fastaPath=\"$fastaFile\" workPath=\"$workDir\" bamreadcount=\"$bamreadcountFile\" bash\n",
    "echo myPythonVariable\\'s value is $fastaPath and $workPath\n",
    "\n",
    "for file in *.loc\n",
    "do\n",
    "\techo $file\n",
    "\tname=${file%.loc}\n",
    "\techo $name\n",
    "\techo \"$name bam-readcount\"\n",
    "\tbam=`ls $workPath*/*/${name}_T_recal.bam`\n",
    "\techo $bam\n",
    "    echo '\\n'\n",
    "\t$bamreadcount -f $fastaPath -w 1 -l $file $bam  > ${name}.readcount &\n",
    "done "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1783.var\n",
      "A9155.var\n",
      "A9878.var\n",
      "AT2550.var\n",
      "AT2822.var\n",
      "AT4415.var\n",
      "AT4808.var\n",
      "C0288.var\n",
      "C0334.var\n",
      "C1572.var\n",
      "C8448.var\n",
      "s1_3076_D.var\n",
      "s1_3295_D.var\n",
      "s1_8588_D.var\n",
      "s2_6088_D.var\n",
      "s2_6339_D.var\n",
      "s2_8811_D.var\n",
      "s3_5832_D.var\n",
      "s3_7299_D.var\n",
      "s3_7979_D.var\n"
     ]
    }
   ],
   "source": [
    "%%script env fpfilter=\"$fpfilterFile\" bamreadcount=\"$bamreadcountFile\"  bash\n",
    "\n",
    "for file in *.var\n",
    "do\n",
    "\techo $file\n",
    "\tperl $fpfilter ${file} ${file%.var}.readcount --output-basename ${file%.var}.fpfilter > ${file%.var}.log &\n",
    "done  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n",
      "[1] 8\n",
      "[1] 9\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 12\n",
      "[1] 13\n",
      "[1] 14\n",
      "[1] 15\n",
      "[1] 16\n",
      "[1] 17\n",
      "[1] 18\n",
      "[1] 19\n",
      "[1] 20\n",
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n",
      "[1] 8\n",
      "[1] 9\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 12\n",
      "[1] 13\n",
      "[1] 14\n",
      "[1] 15\n",
      "[1] 16\n",
      "[1] 17\n",
      "[1] 18\n",
      "[1] 19\n",
      "[1] 20\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "setwd(fpfilterDir)\n",
    "failfiles <- dir(path=\".\",pattern = \"fpfilter.fail\")\n",
    "FPfail <- NULL\n",
    "for(i in 1:length(failfiles)){\n",
    "  print(i)\n",
    "  patient <- unlist(strsplit(failfiles[i],\".\",fixed=TRUE))[1]\n",
    "  tmp <- readLines(failfiles[i])\n",
    "  tmp <- sapply(tmp,function(x){\n",
    "    Tmp <- unlist(strsplit(x,\"\\t\"))\n",
    "    length(Tmp) <- 9\n",
    "    Tmp\n",
    "  })\n",
    "  tmp <- t(tmp)\n",
    "  failID <- paste0(patient,\"-\",tmp[,1],\":\",tmp[,2],\"-\",tmp[,4])\n",
    "  tmp <- cbind(failID,tmp[,9])\n",
    "  rownames(tmp) <- failID\n",
    "  FPfail <- rbind(FPfail,tmp)\n",
    "}\n",
    "\n",
    "passfiles <- dir(path=\".\",pattern = \"fpfilter.pass\")\n",
    "FPpass <- NULL\n",
    "for(i in 1:length(passfiles)){\n",
    "  print(i)\n",
    "  patient <- unlist(strsplit(passfiles[i],\".\",fixed=TRUE))[1]\n",
    "  tmp <- read.delim(passfiles[i],header=FALSE,as.is=TRUE)\n",
    "  tmp <- paste0(patient,\"-\",tmp[,1],\":\",tmp[,2],\"-\",tmp[,4])\n",
    "  FPpass <- c(FPpass,tmp)\n",
    "}\n",
    "\n",
    "length(FPpass) #8967\n",
    "nrow(FPfail) #18886\n",
    "\n",
    "length(FPpass)+nrow(FPfail) #27853\n",
    "\n",
    "all <- c(FPpass,FPfail[,1])\n",
    "table(vcf$fpFid%in%all)\n",
    "#View(vcf[!vcf$fpFid%in%all,])\n",
    "\n",
    "vcf <- data.frame(vcf,FPfilter=\".\",FPfail=\".\",stringsAsFactors=FALSE)\n",
    "vcf[vcf$fpFid%in%FPpass,\"FPfilter\"] <- \"PASS\"\n",
    "vcf[vcf$fpFid%in%rownames(FPfail),\"FPfilter\"] <- \"FAIL\"\n",
    "vcf[vcf$FPfilter==\"FAIL\",\"FPfail\"] <- FPfail[vcf[vcf$FPfilter==\"FAIL\",\"fpFid\"],2]\n",
    "\n",
    "vcf[is.na(vcf$FPfail),\"FPfail\"] <- \"\"\n",
    "vcf[vcf$FPfail==\"\",\"FPfilter\"] <- \"no_readcounts\" # alt absent in the bam readcount file\n",
    "vcf[vcf$FPfail==\"\",\"FPfail\"] <- \".\"\n",
    "vcf[vcf$FPfilter==\".\",\"FPfilter\"] <- \"no_readcounts\"# alt = 0 in the bam readcount file\n",
    "table(vcf$FPfilter)\n",
    "# FAIL no_readcounts          PASS \n",
    "# 15577          3350          8967  \n",
    "rm(FPpass,FPfail,all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# # VirtualNormalCorrection----------\n",
    "# system(\"mkdir VN\")\n",
    "# setwd(\"VN/\")\n",
    "# ID <- 1:nrow(vcf)\n",
    "# VCF <- data.frame(vcf[,1:2],ID,vcf[,3:4],stringsAsFactors = FALSE)\n",
    "# write.table(VCF,sep=\"\\t\",row.names = FALSE,quote=FALSE,file = \"VNinput.vcf\")\n",
    "# n <- ceiling(nrow(VCF)/60)\n",
    "# system(paste0(\"split -l \",n,\" --additional-suffix .vcf VNinput.vcf\"))\n",
    "# \n",
    "# # UNIX\n",
    "# # parallelVN.sh\n",
    "# # parallelVN_combineRes.sh\n",
    "# \n",
    "# VNoutput <- read.delim(\"VNoutput_main.tsv\",as.is=TRUE)\n",
    "# rownames(vcf) <- VCF$ID\n",
    "# vcf <- data.frame(vcf,idNum=VCF$ID,stringsAsFactors = FALSE)\n",
    "# vcf <- data.frame(vcf[VNoutput$xRef,],VNoutput[,9:12],stringsAsFactors = FALSE)\n",
    "# # threshold 5% (~ 22/433) - threshold_highconf 10% (~ 44/433)\n",
    "# table(vcf$VN_occurrences<22&vcf$VN_fullycalled_count>=44)\n",
    "# # FALSE   TRUE \n",
    "# # 220620  84644 \n",
    "# table(vcf$VN_occurrences<3&vcf$VN_fullycalled_count>=110)\n",
    "# # FALSE   TRUE \n",
    "# # 245383  59881 \n",
    "# table(vcf$VN_occurrences<3) # 0.7% (3/433)\n",
    "# # FALSE   TRUE \n",
    "# # 233414  71850 \n",
    "# # table(is.na(vcf$VN_occurrences))\n",
    "# \n",
    "# vcf$vn <- vcf$VN_occurrences<3\n",
    "# table(vcf$vn)\n",
    "# # FALSE   TRUE \n",
    "# # 233414  71850 \n",
    "# \n",
    "# rm(VCF,VNoutput,tmp)\n",
    "\n",
    "#setwd(\"/storage/gluster/vol1/data/PUBLIC/SCAMBIO/ABT414_WES_Analysis/Analysis/\")\n",
    "save(vcf,file = paste(fileDir,\"vcf.rda\",sep=\"\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STEP 2 - ANNOTATION AND DOWNSTREAM ANALYSIS #"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ANNOVAR Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(fileDir) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "ann.in <- vcf[,c(1,2,2,3,4)]\n",
    "colnames(ann.in) <- c(\"Chr\",\"Start\",\"End\",\"Ref\",\"Alt\")\n",
    "\n",
    "cond <- nchar(ann.in[,\"Ref\"])>1&nchar(ann.in[,\"Alt\"])==1\n",
    "ann.in[cond,\"Start\"] <- as.numeric(ann.in[cond,\"Start\"])+1\n",
    "ann.in[cond,\"Ref\"] <- substring(ann.in[cond,\"Ref\"],2,nchar(ann.in[cond,\"Ref\"]))\n",
    "ann.in[cond,\"Alt\"] <- \"-\"\n",
    "ann.in[cond,\"End\"] <- as.numeric(ann.in[cond,\"Start\"])+nchar(ann.in[cond,\"Ref\"])-1\n",
    "\n",
    "cond <- nchar(ann.in[,\"Alt\"])>1&nchar(ann.in[,\"Ref\"])==1\n",
    "ann.in[cond,\"Ref\"] <- \"-\"\n",
    "ann.in[cond,\"Alt\"] <- substring(ann.in[cond,\"Alt\"],2,nchar(ann.in[cond,\"Alt\"]))\n",
    "\n",
    "cond <- nchar(ann.in[,\"Ref\"])>1&nchar(ann.in[,\"Alt\"])>1\n",
    "ann.in[cond,\"End\"] <- as.numeric(ann.in[cond,\"Start\"])+nchar(ann.in[cond,\"Ref\"])-1\n",
    "\n",
    "write.table(ann.in,file=paste(fileDir,\"ann.in\",sep=\"\"),col.names=FALSE,row.names=FALSE,quote=FALSE,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "/storage/gluster/vol1/data/PUBLIC/Tools/annovar2017Jul16/table_annovar.pl ann.in \\\n",
    "/storage/gluster/vol1/data/PUBLIC/Tools/annovar2017Jul16/humandb/ -buildver hg19 -out ann.out \\\n",
    "-remove -protocol refGene,avsnp150,snp138NonFlagged,\\\n",
    "exac03,1000g2015aug_all,esp6500siv2_all,kaviar_20150923,hrcr1,\\\n",
    "cosmic80,clinvar_20170905,\\\n",
    "dbnsfp33a,dbscsnv11,dann,eigen,gerp++gt2,cadd  \\\n",
    "-operation g,f,f,f,f,f,f,f,f,f,f,f,f,f,f,f -nastring . --thread 80"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "ann.out <- read.delim(\"ann.out.hg19_multianno.txt\",as.is=TRUE)\n",
    "paste0(colnames(ann.out),collapse=\"','\")\n",
    "tmp <- c('Chr','Start','End','Ref','Alt','Func.refGene','Gene.refGene','GeneDetail.refGene','ExonicFunc.refGene','AAChange.refGene', #refGene\n",
    "         'avsnp150', #avsnp150\n",
    "         'snp138NonFlagged', #snp138NonFlagged\n",
    "         'ExAC_ALL','ExAC_AFR','ExAC_AMR','ExAC_EAS','ExAC_FIN','ExAC_NFE','ExAC_OTH','ExAC_SAS', #exac03\n",
    "         'X1000g2015aug_all', #1000g2015aug_all\n",
    "         'esp6500siv2_all', #esp6500siv2_all\n",
    "         'Kaviar_AF','Kaviar_AC','Kaviar_AN', #kaviar_20150923\n",
    "         'HRC_AF','HRC_AC','HRC_AN','HRC_non1000G_AF','HRC_non1000G_AC','HRC_non1000G_AN', #hrcr1\n",
    "         'cosmic80', #cosmic80\n",
    "         'CLINSIG','CLNDBN','CLNACC','CLNDSDB','CLNDSDBID', #clinvar_20170905\n",
    "         'SIFT_score','SIFT_converted_rankscore','SIFT_pred', #dbnsfp33a\n",
    "         'Polyphen2_HDIV_score','Polyphen2_HDIV_rankscore','Polyphen2_HDIV_pred', #dbnsfp33a\n",
    "         'Polyphen2_HVAR_score','Polyphen2_HVAR_rankscore','Polyphen2_HVAR_pred', #dbnsfp33a\n",
    "         'LRT_score','LRT_converted_rankscore','LRT_pred', #dbnsfp33a\n",
    "         'MutationTaster_score','MutationTaster_converted_rankscore','MutationTaster_pred', #dbnsfp33a\n",
    "         'MutationAssessor_score','MutationAssessor_score_rankscore','MutationAssessor_pred', #dbnsfp33a\n",
    "         'FATHMM_score','FATHMM_converted_rankscore','FATHMM_pred', #dbnsfp33a\n",
    "         'PROVEAN_score','PROVEAN_converted_rankscore','PROVEAN_pred', #dbnsfp33a\n",
    "         'VEST3_score','VEST3_rankscore','MetaSVM_score','MetaSVM_rankscore','MetaSVM_pred', #dbnsfp33a\n",
    "         'MetaLR_score','MetaLR_rankscore','MetaLR_pred', #dbnsfp33a\n",
    "         'M.CAP_score','M.CAP_rankscore','M.CAP_pred', #dbnsfp33a\n",
    "         'CADD_raw','CADD_raw_rankscore','CADD_phred', #dbnsfp33a\n",
    "         'DANN_score','DANN_rankscore', #dbnsfp33a\n",
    "         'fathmm.MKL_coding_score','fathmm.MKL_coding_rankscore','fathmm.MKL_coding_pred', #dbnsfp33a\n",
    "         'Eigen_coding_or_noncoding','Eigen.raw','Eigen.PC.raw', #dbnsfp33a\n",
    "         'GenoCanyon_score','GenoCanyon_score_rankscore', #dbnsfp33a\n",
    "         'integrated_fitCons_score','integrated_fitCons_score_rankscore','integrated_confidence_value', #dbnsfp33a\n",
    "         'GERP.._RS','GERP.._RS_rankscore', #dbnsfp33a\n",
    "         'phyloP100way_vertebrate','phyloP100way_vertebrate_rankscore','phyloP20way_mammalian','phyloP20way_mammalian_rankscore', #dbnsfp33a\n",
    "         'phastCons100way_vertebrate','phastCons100way_vertebrate_rankscore','phastCons20way_mammalian','phastCons20way_mammalian_rankscore', #dbnsfp33a\n",
    "         'SiPhy_29way_logOdds','SiPhy_29way_logOdds_rankscore', #dbnsfp33a\n",
    "         'Interpro_domain', #dbnsfp33a\n",
    "         'GTEx_V6_gene','GTEx_V6_tissue', #dbnsfp33a\n",
    "         'dbscSNV_ADA_SCORE','dbscSNV_RF_SCORE', #dbscsnv11 SPLICE SITE\n",
    "         'dann', #dann\n",
    "         'Eigen', #eigen\n",
    "         'gerp..gt2', #gerp++gt2\n",
    "         'CADD','CADD_Phred') #cadd\n",
    "\n",
    "sel <- c('Chr','Start','End','Ref','Alt',\n",
    "         'Func.refGene','Gene.refGene','GeneDetail.refGene','ExonicFunc.refGene','AAChange.refGene', 'Interpro_domain',\n",
    "         'avsnp150','snp138NonFlagged',\n",
    "         'ExAC_ALL','X1000g2015aug_all','esp6500siv2_all','Kaviar_AF','HRC_AF',\n",
    "         'cosmic80','CLINSIG','CLNDBN','CLNACC','CLNDSDB','CLNDSDBID',\n",
    "         'SIFT_pred','Polyphen2_HDIV_pred','MutationTaster_pred','PROVEAN_pred',\n",
    "         'dbscSNV_ADA_SCORE','dbscSNV_RF_SCORE',\n",
    "         'dann','Eigen','gerp..gt2','CADD','CADD_Phred')\n",
    "\n",
    "ann.out <- ann.out[,sel]\n",
    "ann.out[ann.out$ExonicFunc.refGene!=\".\",\"Func.refGene\"] <- ann.out[ann.out$ExonicFunc.refGene!=\".\",\"ExonicFunc.refGene\"]\n",
    "ann.out[ann.out$AAChange.refGene!=\".\",\"GeneDetail.refGene\"] <- ann.out[ann.out$AAChange.refGene!=\".\",\"AAChange.refGene\"]\n",
    "ann.out <- ann.out[,!colnames(ann.out)%in%c(\"ExonicFunc.refGene\",\"AAChange.refGene\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final table (MAF-like)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr1.rds\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: doMC\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr2.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr3.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr4.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr5.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr6.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr7.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr8.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr9.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr10.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr11.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr12.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr13.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr14.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr15.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr16.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr17.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr18.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr19.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr20.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr21.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chr22.rds\"\n",
      "[1] \"reading pon file /storage/gluster/vol1/data/PUBLIC/NGS_ReferenceDatabase/firecloud_PoNs/pon_by_chr//pon_chrX.rds\"\n",
      "\n",
      "FALSE  TRUE \n",
      "19207  8676 \n"
     ]
    }
   ],
   "source": [
    "%%R -i ponfilterFile -i ponDir\n",
    "colnames(vcf)\n",
    "sel <- c(\"snpEff\",\"snpEff_func\",\"caller\",\"callerN\",\"t_depth\",\"t_ref_count\", \"t_alt_count\",\n",
    "         \"n_depth\",\"n_ref_count\", \"n_alt_count\", \"t_vaf\",\"n_vaf\",\"FPfilter\",\"FPfail\"#,\n",
    "         #\"VN_occurrences\",\"VN_frequency\",\"VN_fullycalled_count\",\"VN_fullycalled_frequency\",\"vn\"\n",
    ")\n",
    "setdiff(colnames(vcf),sel)\n",
    "setdiff(sel,colnames(vcf))\n",
    "\n",
    "# tumor <- vcf$Tumor_ID\n",
    "# info <- NULL\n",
    "# for(i in 1:length(tumor)){\n",
    "#   print(i)\n",
    "#   tmp <- samples[samples$TUMOR_WES_ID==tumor[i],c(\"Patient_ID\", \"TUMOR_WES_ID\", \"NORMAL_WES_ID\")]\n",
    "#   info <- rbind(info,tmp)\n",
    "# }\n",
    "# colnames(info) <- c(\"Patient_ID\",\"Tumor_ID\",\"Normal_ID\")\n",
    "\n",
    "final <- data.frame(Patient_ID = vcf[, \"Tumor_ID\"], Type= \"SNV\", ann.out, vcf[, sel], stringsAsFactors = FALSE)\n",
    "final[final$Ref==\"-\",\"Type\"] <- \"INS\"\n",
    "final[final$Alt==\"-\",\"Type\"] <- \"DEL\"\n",
    "\n",
    "id <- paste0(final$Patient_ID,\".\",final$Chr,\".\",final$Start,\".\",final$End,\".\",final$Alt)\n",
    "sum(duplicated(id)) #11\n",
    "final <- data.frame(final,duplicated=duplicated(id),stringsAsFactors = FALSE)\n",
    "final <- final[!duplicated(id),]\n",
    "Final <- final\n",
    "rownames(Final) <- NULL\n",
    "# save(Final,file=\"Final.tmp.rda\")\n",
    "\n",
    "# PoNs FILTER (reference_PoNs_final_summed_tokens.hist)----------\n",
    "source(ponfilterFile)\n",
    "Final <- ponFilter(maf=Final, ncores=1, threshold=-2.5, \n",
    "                   pondir=ponDir)\n",
    "table(Final$PON_filter_pass)\n",
    "# FALSE  TRUE \n",
    "# 19207  8676 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GENE DESCRIPTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [====>--------------------------]  17% eta: 20s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [=======>-----------------------]  25% eta: 16s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [=========>---------------------]  33% eta: 13s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [============>------------------]  42% eta: 10s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [===============>---------------]  50% eta: 10s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [=================>-------------]  58% eta: 11s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [====================>----------]  67% eta:  8s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [======================>--------]  75% eta:  6s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [=========================>-----]  83% eta:  4s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [===========================>---]  92% eta:  2s\n",
      "R[write to console]: \n",
      "R[write to console]: Batch submitting query [===============================] 100% eta:  0s\n",
      "                                                                      \n",
      "R[write to console]: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "FALSE  TRUE \n",
      "26926   957 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(biomaRt)\n",
    "ensembl <- useMart(\"ensembl\",dataset=\"hsapiens_gene_ensembl\")\n",
    "genes <- unique(Final$Gene.refGene)\n",
    "# tmp <- listAttributes(ensembl)\n",
    "# tmp[grep(\"description\",tmp$name),]\n",
    "# tmp <- listFilters(ensembl)\n",
    "# tmp[grep(\"external_gene_name\",tmp$name),]\n",
    "info <- getBM(attributes = c(\"external_gene_name\",\"description\"),filters = c(\"external_gene_name\"),values = genes, mart = ensembl)\n",
    "Final <- data.frame(Description=\".\",Final,stringsAsFactors = FALSE)\n",
    "info <- info[!duplicated(info$external_gene_name),]\n",
    "rownames(info) <- info$external_gene_name\n",
    "Final$Description <- info[Final$Gene.refGene,\"description\"]\n",
    "table(is.na(Final$Description))\n",
    "\n",
    "\n",
    "sel <- grep(\";\",Final$Gene.refGene)\n",
    "genes <- Final$Gene.refGene[sel]\n",
    "genes <- sapply(genes,function(x) unlist(strsplit(x,\";\")))\n",
    "genes <- unique(unlist(genes))\n",
    "info <- getBM(attributes = c(\"external_gene_name\",\"description\"),filters = c(\"external_gene_name\"),values = genes, mart = ensembl)\n",
    "description <- sapply(Final$Gene.refGene[sel],function(x){\n",
    "  tmp <- unlist(strsplit(x,\";\"))\n",
    "  if(tmp[1]%in%info$external_gene_name){\n",
    "    a <- info[info$external_gene_name==tmp[1],\"description\"][1]\n",
    "  } else {\n",
    "    a <- \".\"\n",
    "  }\n",
    "  if(tmp[2]%in%info$external_gene_name){\n",
    "    b <- info[info$external_gene_name==tmp[2],\"description\"][1]\n",
    "  } else {\n",
    "    b <- \".\"\n",
    "  }\n",
    "  paste0(a,\" ; \",b)\n",
    "})\n",
    "\n",
    "Final$Description[sel] <- description\n",
    "table(is.na(Final$Description))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # HGVS from SNPEFF----------\n",
    "# snpEff <- sapply(Final$snpEff,function(x){\n",
    "#   temp <- unlist(strsplit(x,\"ANN\"))[2]\n",
    "#   # temp <- unlist(strsplit(temp,\",\"))[1]\n",
    "#   temp <- unlist(strsplit(temp,\",\"))\n",
    "#   if(sum(grepl(\"p.\",temp,fixed=TRUE))>0){\n",
    "#     temp <- temp[grep(\"p.\",temp,fixed=TRUE)][1]\n",
    "#   } else temp <- temp[1]\n",
    "#   info <- unlist(strsplit(temp,\"|\",fixed=TRUE))\n",
    "#   HGVSc <- info[grep(\"c.\",info,fixed=TRUE)][1]\n",
    "#   HGVSp <- info[grep(\"p.\",info,fixed=TRUE)][1]\n",
    "#   return(c(HGVSc,HGVSp))\n",
    "# })\n",
    "# dim(snpEff)\n",
    "# snpEff <- t(snpEff)\n",
    "# colnames(snpEff) <- c(\"HGVSc\",\"HGVSp\")\n",
    "# rownames(snpEff) <- NULL\n",
    "# Final <- data.frame(snpEff,Final,stringsAsFactors = FALSE)\n",
    "# \n",
    "# #----------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotation by variant classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  keep reject \n",
      " 19054   8829 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "Final <- data.frame(effectSnpEff=\".\",effectAnnovar=\".\",Final,stringsAsFactors = FALSE)\n",
    "\n",
    "#snpEff\n",
    "tmp <- unique(Final$snpEff_func)\n",
    "tmp <- unique(unlist(sapply(tmp,function(x) unlist(strsplit(x,\",\")))))\n",
    "tmp <- unique(unlist(sapply(tmp,function(x) unlist(strsplit(x,\"&\")))))\n",
    "sort(tmp)\n",
    "coding <- c(\"conservative_inframe_deletion\",\"conservative_inframe_insertion\",\n",
    "            \"disruptive_inframe_deletion\",\"disruptive_inframe_insertion\",\n",
    "            \"frameshift_variant\",\"initiator_codon_variant\",\"missense_variant\",\n",
    "            \"start_lost\",\"stop_gained\",\"stop_lost\")\n",
    "sort(setdiff(tmp,coding))\n",
    "coding <- sapply(coding,function(x) grep(paste0(\"\\\\b\",x,\"\\\\b\"),Final[,\"snpEff_func\"]))\n",
    "coding <- sort(unique(unlist(coding)))\n",
    "Final[coding,\"effectSnpEff\"] <- \"coding\"\n",
    "splicing <- c(\"splice_acceptor_variant\",\"splice_donor_variant\",\"splice_region_variant\")\n",
    "splicing <- sapply(splicing,function(x) grep(paste0(\"\\\\b\",x,\"\\\\b\"),Final[,\"snpEff_func\"]))\n",
    "splicing <- sort(unique(unlist(splicing)))\n",
    "Final[rownames(Final)%in%splicing & Final$effectSnpEff==\".\",\"effectSnpEff\"] <- \"splicing\"\n",
    "Final$effectSnpEff[Final$effectSnpEff==\".\"] <- \"noncoding\"\n",
    "table(Final$effectSnpEff)\n",
    "# coding noncoding  splicing \n",
    "# 13201      8874      5808 \n",
    "#Annovar\n",
    "tmp <- unique(Final[,\"Func.refGene\"])\n",
    "tmp <- sort(tmp)\n",
    "paste0(tmp,collapse=\"','\")\n",
    "tmp <- c('downstream','frameshift deletion','frameshift insertion','intergenic','intronic',\n",
    "         'ncRNA_exonic','ncRNA_exonic;splicing','ncRNA_intronic','ncRNA_splicing',\n",
    "         'nonframeshift deletion','nonframeshift insertion','nonsynonymous SNV','splicing',\n",
    "         'stopgain','stoploss','synonymous SNV','unknown','upstream','upstream;downstream',\n",
    "         'UTR3','UTR5','UTR5;UTR3')\n",
    "Final[Final$Func.refGene%in%c('frameshift deletion','frameshift insertion','frameshift substitution',\n",
    "                              'nonframeshift deletion','nonframeshift insertion','nonframeshift substitution',\n",
    "                              'nonsynonymous SNV','stopgain','stoploss'),\"effectAnnovar\"] <- \"coding\" \n",
    "Final[Final$Func.refGene%in%c('splicing','ncRNA_exonic;splicing'),\"effectAnnovar\"] <- 'splicing' \n",
    "Final[Final$effectAnnovar==\".\",\"effectAnnovar\"] <- \"noncoding\"\n",
    "table(Final$effectAnnovar)\n",
    "# \n",
    "# coding noncoding  splicing \n",
    "# 11149     16450       284 \n",
    "\n",
    "Final <- data.frame(effectFilt=\"reject\",Final,stringsAsFactors = FALSE)\n",
    "Final[Final$effectSnpEff%in%c(\"coding\",\"splicing\")|Final$effectAnnovar%in%c(\"coding\",\"splicing\"),\"effectFilt\"] <- \"keep\"\n",
    "table(Final$effectFilt)\n",
    "# keep reject \n",
    "# 19054   8829 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotation of gene paralogs with high homology (http://massgenomics.org/2013/06/ngs-false-positives.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     . reject \n",
      " 22994   4889 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "Final <- data.frame(blackGenes=\".\",Final,stringsAsFactors = FALSE)\n",
    "artifacts <- c(\"LOC\",\"ENS\",\"FAM\",\"GOL\",\"PRA\",\"NBP\",\"POT\",\"DEF\",\"MUC\",\"KRT\",\"WAS\",\"ANK\",\"TRI\",\"FRG\",\"HLA\",paste0(\"OR\",1:9))\n",
    "# artifacts <- c(\"LOC\",\"ENS\",\"FAM\",\"GOL\",\"PRA\",\"NBP\",\"DEF\",\"MUC\",\"KRT\",\"WAS\",\"ANK\",\"FRG\",\"HLA\",paste0(\"OR\",1:9)) #KEEP POT AND TRIM\n",
    "genesSuff <- sapply(Final[,\"Gene.refGene\"],function(x) substring(unlist(strsplit(x,\",\"))[1],1,3))\n",
    "toRemove <- genesSuff%in%artifacts\n",
    "Final[toRemove,\"blackGenes\"] <- \"reject\"\n",
    "genesSuff <- sapply(Final[,\"Gene.refGene\"],function(x) substring(unlist(strsplit(x,\",\"))[1],1,4))\n",
    "toRemove <- genesSuff%in%c(\"PLIN\",\"CELA\",\"SRA1\")\n",
    "Final[toRemove,\"blackGenes\"] <- \"reject\"\n",
    "artifacts <- c(\"ATXN1\",\"PBRM1\",\"ZNF814\",\"MSH3\",\"TTN\",\"USH2A\")\n",
    "genes <- sapply(Final[,\"Gene.refGene\"],function(x) unlist(strsplit(x,\",\"))[1])\n",
    "toRemove <- genes%in%artifacts\n",
    "Final[toRemove,\"blackGenes\"] <- \"reject\"\n",
    "table(Final$blackGenes)\n",
    "# . reject \n",
    "# 22994   4889 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ExAC, 1000Genomes, esp6500siv2_all, Kaviar_AF, HRC_AF allele frequency < 5% annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     . reject \n",
      " 20458   7425 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "Final <- data.frame(SNP=\".\",Final,stringsAsFactors = FALSE)\n",
    "toReject <- as.numeric(Final[,\"ExAC_ALL\"])>=0.05\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "table(toReject)\n",
    "# FALSE  TRUE \n",
    "# 21888  5995  \n",
    "Final[toReject,\"SNP\"] <- \"reject\"\n",
    "toReject <- as.numeric(Final[,\"X1000g2015aug_all\"])>=0.05\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "table(toReject)\n",
    "# FALSE   TRUE \n",
    "# 25356  2527 \n",
    "Final[toReject,\"SNP\"] <- \"reject\"\n",
    "toReject <- as.numeric(Final[,\"esp6500siv2_all\"])>=0.05\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "table(toReject)\n",
    "# FALSE   TRUE \n",
    "# 26081  1802\n",
    "Final[toReject,\"SNP\"] <- \"reject\"\n",
    "toReject <- as.numeric(Final[,\"Kaviar_AF\"])>=0.05\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "table(toReject)\n",
    "# FALSE   TRUE \n",
    "# 26529  1354 \n",
    "Final[toReject,\"SNP\"] <- \"reject\"\n",
    "toReject <- as.numeric(Final[,\"HRC_AF\"])>=0.05\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "table(toReject)\n",
    "# FALSE   TRUE \n",
    "# 27154   729 \n",
    "Final[toReject,\"SNP\"] <- \"reject\"\n",
    "table(Final$SNP)\n",
    "# . reject \n",
    "#  20729   7154  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# snp138NonFlagged annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     . reject \n",
      " 15791  12092 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "sum(toRemove <- Final[,\"snp138NonFlagged\"]!=\".\") #8935\n",
    "Final[toRemove,\"SNP\"] <- \"reject\"\n",
    "table(Final$SNP)\n",
    "# . reject \n",
    "# 15986  11897 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 50-mer alignability annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "Final <- data.frame(fiftymer=\".\",Final,stringsAsFactors = FALSE)\n",
    "bed <- Final[,c(\"Chr\",\"Start\",\"End\")]\n",
    "write.table(bed,file='Final.bed',col.names=FALSE,row.names=FALSE,quote=FALSE,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"BEDTOOLS_DIR\"] = \"/glusterfs/biostorage1/brick1/bcbio/anaconda/pkgs/bedtools-2.27.1-1/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script env merPath=\"$merFile\" merScore=\"$merScore\" bedtools=\"$BEDTOOLS_PATH\" bash\n",
    "$bedtools intersect -wo -a Final.bed -b $merPath > $merScore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     . reject \n",
      " 22962   4921 \n"
     ]
    }
   ],
   "source": [
    "%%R -i merScore \n",
    "score <- read.delim(merScore ,header=FALSE,as.is=TRUE)\n",
    "sum(score[,1]=='')\n",
    "# 0\n",
    "# score <- score[!score[,1]=='',c(1,2,3,7)]\n",
    "score <- score[,c(1,2,3,7)]\n",
    "id <- paste0(score[,1],\"-\",score[,2])\n",
    "sum(duplicated(id))\n",
    "score <- score[!duplicated(id),]\n",
    "id <- paste0(score[,1],\"-\",score[,2])\n",
    "rownames(score) <- id\n",
    "ID <- paste0(bed[,1],\"-\",bed[,2])\n",
    "Score <- score[ID,4]\n",
    "Final <- cbind(Score,Final)\n",
    "sum(is.na(Score))\n",
    "#0\n",
    "# View(Final[is.na(Score),]) #strange contigs (ie: chr1_gl000192_random)\n",
    "toRemove <- which(as.numeric(Final[,'Score'])<0.3)\n",
    "Final[toRemove,\"fiftymer\"] <- \"reject\"\n",
    "table(Final$fiftymer)\n",
    "# . reject \n",
    "# 22962   4921 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Masked region annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "bed <- Final[,c(\"Chr\",\"Start\",\"End\")]\n",
    "write.table(bed,file='Finalmask.bed',col.names=FALSE,row.names=FALSE,quote=FALSE,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script env maskPath=\"$maskFile\" maskScore=\"$maskScore\" bedtools=\"$BEDTOOLS_PATH\" bash\n",
    "$bedtools intersect -wao -a Finalmask.bed -b $maskPath > $maskScore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " fail  pass \n",
      " 5656 22227 \n"
     ]
    }
   ],
   "source": [
    "%%R -i maskScore\n",
    "score <- read.delim(maskScore,header=FALSE,as.is=TRUE)\n",
    "id <- paste0(score[,1],\":\",score[,2],\"-\",score[,3])\n",
    "table(duplicated(id))\n",
    "score <- score[!duplicated(id),]\n",
    "id <- paste0(score[,1],\":\",score[,2],\"-\",score[,3])\n",
    "rownames(score) <- id\n",
    "ID <- paste0(bed[,1],\":\",bed[,2],\"-\",bed[,3])\n",
    "table(ID%in%id)\n",
    "Score <- score[ID,7]\n",
    "Final <- data.frame(MaskedFilter=Score,Final,stringsAsFactors=FALSE)\n",
    "Final[Final[,\"MaskedFilter\"]==\"pilot\",\"MaskedFilter\"] <- \"pass\"\n",
    "Final[Final[,\"MaskedFilter\"]==\".\",\"MaskedFilter\"] <- \"fail\"\n",
    "table(Final$MaskedFilter)\n",
    "# fail   pass \n",
    "#  5656 22227 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FLAGS genes annotation (https://bmcmedgenomics.biomedcentral.com/articles/10.1186/s12920-014-0064-y#Sec29)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i flagsFile\n",
    "Final <- data.frame(FLAGS=\".\",FLAGSscore=0,Final,stringsAsFactors = FALSE)\n",
    "FLAGS <- read.delim(flagsFile,sep=\"\\t\",as.is=TRUE,header = FALSE)\n",
    "# quantile(FLAGS[,2])\n",
    "# # 0%  25%  50%  75% 100% \n",
    "# # 1   15   28   49 2659 \n",
    "# quantile(FLAGS[,2],0.99) #195\n",
    "# View(FLAGS[FLAGS[,2]>195,])\n",
    "tmp <- FLAGS$V1\n",
    "# sum(duplicated(tmp))\n",
    "# tmp[duplicated(tmp)]\n",
    "# # \"01-Mar\" \"02-Mar\"\n",
    "# View(FLAGS[order(tmp),])\n",
    "FLAGS <- FLAGS[!duplicated(tmp),]\n",
    "rownames(FLAGS) <- FLAGS$V1\n",
    "genes <- sapply(Final[,\"Gene.refGene\"],function(x) unlist(strsplit(x,\",\"))[1])\n",
    "Final$FLAGSscore <- FLAGS[genes,2]\n",
    "table(is.na(Final$FLAGSscore))\n",
    "Final$FLAGSscore[is.na(Final$FLAGSscore)] <- 0\n",
    "# View(Final[order(Final[,1],decreasing=TRUE),])\n",
    "sum(Final[,\"FLAGSscore\"]>195) #2606\n",
    "Final[Final[,\"FLAGSscore\"]>195,\"FLAGS\"] <- \"reject\"\n",
    "table(Final$FLAGS)\n",
    "# . reject \n",
    "# 25277   2606\n",
    "save(Final,file=\"Final.tmp.rda\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cBio Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "bed <- Final[,c(\"Chr\",\"Start\",\"End\")]\n",
    "write.table(bed,file=\"SomaticVariants.bed\",col.names=FALSE,row.names=FALSE,quote=FALSE,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script env cbioPath=\"$cbioFile\" cbioScore=\"$cbioScore\" bedtools=\"$BEDTOOLS_PATH\" bash\n",
    "$bedtools intersect -wao -f 1 -a SomaticVariants.bed -b $cbioPath > $cbioScore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading objects:\n",
      "  cBioMutDb\n"
     ]
    }
   ],
   "source": [
    "%%R -i cbioDB -i cbioScore\n",
    "selVar_cBio <- read.delim(cbioScore,header=FALSE,as.is=TRUE)\n",
    "load(cbioDB,verbose=TRUE)\n",
    "ID <- paste0(selVar_cBio[,1],\":\",selVar_cBio[,2],\"-\",selVar_cBio[,3])\n",
    "id <- paste0(Final[,\"Chr\"],\":\",Final[,\"Start\"],\"-\",Final[,\"End\"])\n",
    "\n",
    "cBio <- function(x){\n",
    "  cBioID <- as.numeric(sub(\"cBio_\",\"\",unique(selVar_cBio[ID==x,7])))\n",
    "  a <- paste0(cBioID,collapse=\"|\")\n",
    "  b <- paste0(cBioMutDb[cBioID,\"AAchange\"],collapse=\"|\")\n",
    "  c <- paste0(cBioMutDb[cBioID,\"KnownVar\"],collapse=\"|\")\n",
    "  d <- paste0(cBioMutDb[cBioID,\"StudyID\"],collapse=\"|\")\n",
    "  e <- sum(as.numeric(cBioMutDb[cBioID,\"SamplesNum\"]))\n",
    "  f <- paste0(cBioMutDb[cBioID,\"Freq\"],collapse=\"|\")\n",
    "  e <- c(a,b,c,d,e,f)\n",
    "  return(e)\n",
    "}\n",
    "\n",
    "cl <- makePSOCKcluster(20,outfile=\"\")\n",
    "registerDoParallel(cl)\n",
    "#clusterExport(cl, c(\"cBioMutDb\",\"id\"))\n",
    "cBioAnn <- foreach(i = 1:length(id)) %dopar% {\n",
    "  tmp <- cBio(id[i])\n",
    "  return(tmp)\n",
    "}\n",
    "stopCluster(cl)\n",
    "\n",
    "cBioAnn <- do.call(rbind,cBioAnn)\n",
    "nrow(cBioAnn) #27883\n",
    "colnames(cBioAnn) <- c(\"cBioId\",\"cBioAAchange\",\"cBioKnownVar\",\"cBioStudyId\",\"cBioSamplesNum\",\"cBioFreq\")\n",
    "Final <- data.frame(Final,cBioAnn,stringsAsFactors = FALSE)\n",
    "# View(Final[cBioAnn[,1]!=\"NA\",])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Onco-annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i censusFile -i natureFile -i oncoKBFile\n",
    "census <- read.delim(censusFile,as.is=TRUE)\n",
    "nature <- read.delim(natureFile,as.is=TRUE)\n",
    "oncoKB <- read.delim(oncoKBFile,as.is=TRUE)\n",
    "oncoKB <- oncoKB[oncoKB$Oncogenicity==\"Oncogenic\",]\n",
    "\n",
    "onco <- function(i){\n",
    "  x <- Final[i,]\n",
    "  gene <- unlist(strsplit(as.character(x[\"Gene.refGene\"]),\";\"))[1]\n",
    "  #census\n",
    "  if(gene%in%census$Gene.Symbol){\n",
    "    a <- census[census$Gene.Symbol%in%gene,\"Role.in.Cancer\"]\n",
    "    b <- paste0(census[census$Gene.Symbol%in%gene,c(\"Tumour.Types.Germline.\",\"Cancer.Syndrome\")],collapse=\"|\")\n",
    "  } else {\n",
    "    a <- \".\"\n",
    "    b <- \".\"\n",
    "  }\n",
    "  #nature\n",
    "  if(gene%in%nature$Gene..Symbol){\n",
    "    c <- paste0(nature[nature$Gene..Symbol%in%gene,\"Mechanism.of.action.of.CPG.mutations\"],collapse=\"|\")\n",
    "    d <- nature[nature$Gene..Symbol%in%gene,\"Cancer.syndrome.s.\"]\n",
    "  } else {\n",
    "    c <- \".\"\n",
    "    d <- \".\"\n",
    "  }\n",
    "  #oncoKB\n",
    "  if(gene%in%oncoKB$Gene) {\n",
    "    e <- paste0(oncoKB[oncoKB$Gene%in%gene,\"Mutation.Effect\"],collapse=\"|\")\n",
    "    f <- paste0(oncoKB[oncoKB$Gene%in%gene,\"Alteration\"],collapse=\"|\")\n",
    "  } else {\n",
    "    e <- \".\"\n",
    "    f <- \".\"\n",
    "  }\n",
    "  g <- c(a,b,c,d,e,f)\n",
    "  return(g)\n",
    "}\n",
    "\n",
    "cl <- makePSOCKcluster(30,outfile=\"\")\n",
    "registerDoParallel(cl)\n",
    "Onco <- foreach(i = 1:nrow(Final)) %do% {\n",
    "  tmp <- onco(i)\n",
    "  return(tmp)\n",
    "}\n",
    "stopCluster(cl)\n",
    "\n",
    "Onco <- do.call(rbind,Onco)\n",
    "nrow(Onco) #27883\n",
    "colnames(Onco) <- c(\"census\",\"censusGermline\",\"nature\",\"natureGermline\",\"oncoKB\",\"oncoKBmut\")\n",
    "\n",
    "Final <- data.frame(Final,Onco ,stringsAsFactors = FALSE)\n",
    "#save(Final,file=\"Final.tmp.rda\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Panglioma recurrent mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i smgFile\n",
    "smg <- read.csv(smgFile,as.is=TRUE)\n",
    "genes <- sapply(Final[,\"Gene.refGene\"],function(x) unlist(strsplit(x,\",\"))[1])\n",
    "smgOcc <- NULL\n",
    "for(i in 1:nrow(Final)){\n",
    "  if(genes[i]%in%smg[,1]){\n",
    "    smgOcc <- c(smgOcc,smg[smg[,1]==genes[i],2])\n",
    "  }\n",
    "  else smgOcc <- c(smgOcc,0)\n",
    "}\n",
    "Final <- data.frame(Final,smgOcc,stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Damaging prediction counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                  .          frameshift             inframe          pathogenic \n",
      "              16427                 519                1349                  23 \n",
      "pathogenic missense            splicing            stoploss  tolerated missense \n",
      "               3549                 284                   7                5400 \n",
      "         truncating \n",
      "                325 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# pred <- Final[,colnames(Final)[grep(\"_pred\",colnames(Final))]]\n",
    "# pred <- pred[,c(\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")]\n",
    "unique(Final$SIFT_pred) # \"D\" \".\" \"T\"\n",
    "unique(Final$Polyphen2_HDIV_pred) #\"P\" \".\" \"B\" \"D\"\n",
    "Final$Polyphen2_HDIV_pred[Final$Polyphen2_HDIV_pred==\"P\"] <- \"D\"\n",
    "Final$Polyphen2_HDIV_pred[Final$Polyphen2_HDIV_pred==\"B\"] <- \"T\"\n",
    "unique(Final$MutationTaster_pred) #\"D\" \".\" \"N\" \"A\" \"P\"\n",
    "Final$MutationTaster_pred[Final$MutationTaster_pred==\"A\"] <- \"D\"\n",
    "Final$MutationTaster_pred[Final$MutationTaster_pred==\"N\"] <- \"T\"\n",
    "Final$MutationTaster_pred[Final$MutationTaster_pred==\"P\"] <- \"T\"\n",
    "unique(Final$PROVEAN_pred) #\"N\" \".\" \"D\"\n",
    "Final$PROVEAN_pred[Final$PROVEAN_pred==\"N\"] <- \"T\"\n",
    "tmp <- apply(Final[,c(\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")],1,function(x){\n",
    "  d <- sum(x%in%c(\"D\"))\n",
    "  t <- sum(x%in%c(\"T\"))\n",
    "  u <- sum(x%in%c(\".\"))\n",
    "  c(d,t,u)\n",
    "})\n",
    "tmp <- t(tmp)\n",
    "colnames(tmp) <- c(\"Damaging\",\"Tolerated\",\"Unpredicted\")\n",
    "\n",
    "Final <- data.frame(Pathogenicity=\".\",MissenseDamaging=tmp[,1],Final,stringsAsFactors = FALSE)\n",
    "\n",
    "table(Final$Func.refGene)\n",
    "\n",
    "Final[Final$Func.refGene==\"nonsynonymous SNV\"&as.numeric(Final$MissenseDamaging)>=2,\"Pathogenicity\"] <- \"pathogenic missense\"\n",
    "Final[Final$Func.refGene==\"nonsynonymous SNV\"&as.numeric(Final$MissenseDamaging)<2,\"Pathogenicity\"] <- \"tolerated missense\"\n",
    "\n",
    "Final[Final$Pathogenicity==\".\"&as.numeric(Final$MissenseDamaging)>=2,\"Pathogenicity\"] <- \"pathogenic\"\n",
    "\n",
    "Final[Final$Func.refGene%in%c(\"frameshift deletion\",\"frameshift insertion\",\"frameshift substitution\"),\"Pathogenicity\"] <- \"frameshift\"\n",
    "Final[Final$Func.refGene%in%c(\"frameshift deletion\",\"frameshift insertion\",\"frameshift substitution\"),\n",
    "      c(\"MissenseDamaging\",\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")] <- \".\"\n",
    "\n",
    "Final[Final$Func.refGene%in%c(\"splicing\",\"exonic;splicing\"),\"Pathogenicity\"] <- \"splicing\"\n",
    "Final[Final$Func.refGene%in%c(\"splicing\",\"exonic;splicing\"),\n",
    "      c(\"MissenseDamaging\",\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")] <- \".\"\n",
    "\n",
    "Final[Final$Func.refGene%in%c(\"stopgain\"),\"Pathogenicity\"] <- \"truncating\"\n",
    "Final[Final$Func.refGene%in%c(\"stopgain\"),\n",
    "      c(\"MissenseDamaging\",\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")] <- \".\"\n",
    "\n",
    "Final[Final$Func.refGene%in%c(\"stoploss\"),\"Pathogenicity\"] <- \"stoploss\"\n",
    "Final[Final$Func.refGene%in%c(\"stoploss\"),\n",
    "      c(\"MissenseDamaging\",\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")] <- \".\"\n",
    "\n",
    "Final[Final$Func.refGene%in%c(\"nonframeshift deletion\",\"nonframeshift insertion\",\"nonframeshift substitution\"),\"Pathogenicity\"] <- \"inframe\"\n",
    "Final[Final$Func.refGene%in%c(\"nonframeshift deletion\",\"nonframeshift insertion\",\"nonframeshift substitution\"),\n",
    "      c(\"MissenseDamaging\",\"SIFT_pred\",\"Polyphen2_HDIV_pred\",\"MutationTaster_pred\",\"PROVEAN_pred\")] <- \".\"\n",
    "\n",
    "table(Final$Pathogenicity)\n",
    "# .          frameshift             inframe          pathogenic pathogenic missense            splicing            stoploss  tolerated missense \n",
    "# 16427                 519                1349                  23                3549                 284                   7                5400 \n",
    "# truncating \n",
    "# 325 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DOWNSTREAM SELECTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "     . reject \n",
      " 12495  15388 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# 1000Genomes and dbNSFP filter AF â‰¤ 0.0004\n",
    "Final <- data.frame(PopFreq=\".\",Final,stringsAsFactors = FALSE)\n",
    "toReject <- as.numeric(Final[,\"ExAC_ALL\"])>0.0004\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "Final[toReject,\"PopFreq\"] <- \"reject\"\n",
    "toReject <- as.numeric(Final[,\"X1000g2015aug_all\"])>0.0004\n",
    "toReject[is.na(toReject)] <- FALSE\n",
    "Final[toReject,\"PopFreq\"] <- \"reject\"\n",
    "table(Final$PopFreq)\n",
    "# . reject \n",
    "# 15887  11996  \n",
    "\n",
    "# Low confidence n_alt_count > 1 | t_depth < 15 | t_alt_count <= 3 | T_vaf < 0.05\n",
    "Final <- data.frame(LowConf=\".\",Final,stringsAsFactors = FALSE)\n",
    "\n",
    "sum(sel <- is.na(Final$n_vaf))\n",
    "unique(Final[sel,\"Patient_ID\"])\n",
    "# sum(sel <- grepl(\"NA\",Final$n_vaf))\n",
    "# unique(Final[sel,\"caller\"])\n",
    "\n",
    "lowConf <- function(i){\n",
    "  if(!is.na(Final$n_vaf[i])){\n",
    "    tmp <- Final$n_alt_count[i]\n",
    "    # tmp <- as.numeric(unlist(strsplit(Final$n_alt_count[i],\"|\",fixed=TRUE)))\n",
    "    # if(length(tmp)>1){\n",
    "    #   toRej1 <- sum(tmp>1)>1\n",
    "    # } else {\n",
    "    toRej1 <- as.numeric(tmp)>1\n",
    "    # } \n",
    "  } else toRej1 <- FALSE\n",
    "  tmp <- Final$t_depth[i]\n",
    "  # tmp <- as.numeric(unlist(strsplit(Final$t_depth[i],\"|\",fixed=TRUE)))\n",
    "  # if(length(tmp)>1) {\n",
    "  #   toRej2 <- sum(tmp<15)>1\n",
    "  # } else {\n",
    "  toRej2 <- tmp<15\n",
    "  # }\n",
    "  tmp <- Final$t_alt_count[i]\n",
    "  # tmp <- as.numeric(unlist(strsplit(Final$t_alt_count[i],\"|\",fixed=TRUE)))\n",
    "  # if(length(tmp)>1) {\n",
    "  #   toRej3 <- sum(tmp<=3)>1\n",
    "  # } else {\n",
    "  toRej3 <- tmp<=3\n",
    "  # }\n",
    "  tmp <- Final$t_vaf[i]\n",
    "  # tmp <- as.numeric(unlist(strsplit(Final$t_vaf[i],\"|\",fixed=TRUE)))\n",
    "  # if(length(tmp)>1) {\n",
    "  #   toRej4 <- sum(tmp<0.05)>1\n",
    "  # } else {\n",
    "  toRej4 <- tmp<=0.05\n",
    "  # }\n",
    "  toRej <- toRej1|toRej2|toRej3|toRej4\n",
    "  return(toRej)\n",
    "}\n",
    "\n",
    "# res <- NULL\n",
    "# for(i in 1:nrow(Final)){\n",
    "#   print(i)\n",
    "#   tmp <- lowConf(i)\n",
    "#   res <- c(res,tmp)\n",
    "# }\n",
    "\n",
    "\n",
    "cl <- makePSOCKcluster(30,outfile=\"\")\n",
    "registerDoParallel(cl)\n",
    "res <- foreach(i = 1:nrow(Final)) %dopar% {\n",
    "  tmp <- lowConf(i)\n",
    "  return(tmp)\n",
    "}\n",
    "stopCluster(cl)\n",
    "res <- unlist(res)\n",
    "\n",
    "table(res)\n",
    "# FALSE   TRUE \n",
    "# 12495 15388 \n",
    "# table(is.na(res))\n",
    "Final[res,\"LowConf\"] <- \"reject\"\n",
    "table(Final$LowConf)\n",
    "# . reject \n",
    "#  12495  15388 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TABLE FINALIZATION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# tmp <- paste0(colnames(Final),collapse = \"','\")\n",
    "# \n",
    "# tmp <- c('Description','LowConf','PopFreq','Pathogenicity','MissenseDamaging','FLAGSscore','FLAGS',\n",
    "#          'MaskedFilter','Score','fiftymer','SNP','blackGenes','effectFilt','effectSnpEff','effectAnnovar','Patient_ID',\n",
    "#          'Type','Chr','Start','End','Ref','Alt','Func.refGene','Gene.refGene','GeneDetail.refGene','Interpro_domain',\n",
    "#          'avsnp150','snp138NonFlagged','ExAC_ALL','X1000g2015aug_all','esp6500siv2_all','Kaviar_AF','HRC_AF','cosmic80','CLINSIG',\n",
    "#          'CLNDBN','CLNACC','CLNDSDB','CLNDSDBID','SIFT_pred','Polyphen2_HDIV_pred','MutationTaster_pred','PROVEAN_pred',\n",
    "#          'dbscSNV_ADA_SCORE','dbscSNV_RF_SCORE','dann','Eigen','gerp..gt2','CADD','CADD_Phred',\n",
    "#          'snpEff','snpEff_func','caller','t_depth','t_ref_count','t_alt_count','n_depth','n_ref_count','n_alt_count','t_vaf','n_vaf',\n",
    "#          'FPfilter','FPfail','cBio','cBioId','cBioAAchange','cBioKnownVar','cBioStudyId','cBioSamplesNum','cBioFreq',\n",
    "#          'census','censusGermline','nature','natureGermline','oncoKB','oncoKBmut')\n",
    "\n",
    "tmp <- colnames(Final)\n",
    "\n",
    "sel <- c(\"Patient_ID\",#'Tumor_ID',\"Normal_ID\", #SAMPLE\n",
    "         'caller','callerN',\"duplicated\",'Type','Chr','Start','End','Ref','Alt',\n",
    "         #\"HGVSc\",\"HGVSp\", #VARIANT\n",
    "         'Gene.refGene','Description','Func.refGene','GeneDetail.refGene','snpEff','snpEff_func','Interpro_domain','effectFilt','effectSnpEff','effectAnnovar', #GENE ANNOTATION\n",
    "         'SNP','PopFreq','avsnp150','snp138NonFlagged','ExAC_ALL','X1000g2015aug_all','esp6500siv2_all','Kaviar_AF','HRC_AF', #POPULATION FREQUENICIES\n",
    "         'FPfilter','FPfail','MaskedFilter','Score','fiftymer','FLAGSscore','FLAGS','blackGenes', #FALSE POSITIVE FILTERS\n",
    "         #\"VN_occurrences\",\"VN_frequency\",\"VN_fullycalled_count\",\"VN_fullycalled_frequency\",\"vn\",\n",
    "         \"PON_filter\",\"PON_filter_pass\", #FALSE POSITIVE FILTERS\n",
    "         'LowConf','t_depth','t_ref_count','t_alt_count','n_depth','n_ref_count','n_alt_count','t_vaf','n_vaf', #READ COUNTS\n",
    "         'Pathogenicity','MissenseDamaging','SIFT_pred','Polyphen2_HDIV_pred','MutationTaster_pred','PROVEAN_pred', #PATHOGENICITY\n",
    "         'dbscSNV_ADA_SCORE','dbscSNV_RF_SCORE','dann','Eigen','gerp..gt2','CADD','CADD_Phred', #WGS PATHOGENICITY\n",
    "         'cosmic80','CLINSIG','CLNDBN','CLNACC','CLNDSDB','CLNDSDBID', #CLINICAL ANNOTATION\n",
    "         'census','censusGermline','nature','natureGermline','oncoKB','oncoKBmut',\"smgOcc\", #ONCO ANNOTATION\n",
    "         'cBioId','cBioAAchange','cBioKnownVar','cBioStudyId','cBioSamplesNum','cBioFreq' #cBio ANNOTATION\n",
    ")\n",
    "setdiff(tmp,sel)\n",
    "setdiff(sel,tmp)\n",
    "Final <- Final[,sel]\n",
    "\n",
    "Final <- Final[order(as.numeric(Final$Start)),]\n",
    "Final <- Final[order(Final$Gene.refGene),]\n",
    "#Final <- Final[order(Final$Tumor_ID),]\n",
    "rownames(Final) <- NULL\n",
    "# tmp <- sapply(Final$Patient_ID, function(x) unlist(strsplit(x,\"_\"))[1])\n",
    "# Final <- Final[order(tmp),]\n",
    "\n",
    "Final[Final==\"\"] <- \".\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HGVS from SNPEFF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "snpEff <- sapply(Final$snpEff,function(x){\n",
    "  temp <- unlist(strsplit(x,\"ANN\"))[2]\n",
    "  # temp <- unlist(strsplit(temp,\",\"))[1]\n",
    "  temp <- unlist(strsplit(temp,\",\"))\n",
    "  if(sum(grepl(\"p.\",temp,fixed=TRUE))>0){\n",
    "    temp <- temp[grep(\"p.\",temp,fixed=TRUE)][1]\n",
    "  } else temp <- temp[1]\n",
    "  info <- unlist(strsplit(temp,\"|\",fixed=TRUE))\n",
    "  HGVSc <- info[grep(\"c.\",info,fixed=TRUE)][1]\n",
    "  HGVSp <- info[grep(\"p.\",info,fixed=TRUE)][1]\n",
    "  return(c(HGVSc,HGVSp))\n",
    "})\n",
    "dim(snpEff)\n",
    "snpEff <- t(snpEff)\n",
    "colnames(snpEff) <- c(\"HGVSc\",\"HGVSp\")\n",
    "rownames(snpEff) <- NULL\n",
    "Final <- data.frame(snpEff,Final,stringsAsFactors = FALSE)\n",
    "# Final[,c(\"HGVSc\",\"HGVSp\")] <- snpEff[,c(\"HGVSc\",\"HGVSp\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "Final <- data.frame(selected=\".\",Final,stringsAsFactors = FALSE)\n",
    "Final$selected[grepl(\"path\",Final$CLINSIG,ignore.case = TRUE)] <- \"yes\"\n",
    "Final$selected[Final$Gene.refGene%in%c(\"NF1\",\"SUZ12\",\"EED\")] <- \"yes\"\n",
    "table(Final$selected) # 36 \n",
    "# View(Final[Final$selected==\"yes\",])\n",
    "reject <- Final$effectFilt==\"reject\" |  Final$SNP==\"reject\" | Final$fiftymer==\"reject\" | Final$FLAGS==\"reject\" | Final$blackGenes==\"reject\"\n",
    "# View(Final[reject&Final$selected==\"yes\",])\n",
    "table(reject&Final$selected==\"yes\") #3\n",
    "Final$selected[reject&Final$selected==\".\"] <- \"no\"\n",
    "table(Final$selected)\n",
    "# .    no   yes \n",
    "# 8591 19261    36 \n",
    "\n",
    "FinalS <- Final[Final$selected!=\"no\",]\n",
    "rownames(FinalS) <- NULL\n",
    "\n",
    "table(FP_1 <- FinalS$FPfilter==\"FAIL\" & FinalS$MaskedFilter==\"fail\") #472\n",
    "table(FP_tmp <- FinalS$PopFreq==\"reject\" | FinalS$FPfilter==\"FAIL\" | FinalS$MaskedFilter==\"fail\" | FinalS$LowConf==\"reject\") #5707\n",
    "table(FP_2 <- FP_tmp & (FinalS$PON_filter_pass==FALSE | FinalS$callerN<3)) #5329\n",
    "table(FP_3 <- FinalS$effectAnnovar==\"noncoding\") #2414\n",
    "table(FP_4 <- FinalS$FPfilter==\"no_readcounts\"&FinalS$callerN<3&(nchar(FinalS$Ref)>3|nchar(FinalS$Alt)>3)) #155\n",
    "table(FP_5 <- FinalS$callerN<3 & FinalS$PON_filter_pass==FALSE) #3565 \n",
    "#table(FP_6 <-  (FinalS$PON_filter > -10 | (FinalS$avsnp150!=\".\" & FinalS$cosmic80==\".\"))) #5442  \n",
    "\n",
    "# tmp <- FinalS[!(FP_1 | FP_2 | FP_3 | FP_4 | FP_5) & FP_6,]\n",
    "# View(tmp)\n",
    "\n",
    "table(FinalS$toCheck <- (FP_1 | FP_2 | FP_3 | FP_4 | FP_5 ) & grepl(\"path\",FinalS$CLINSIG,ignore.case = TRUE)) #11 \n",
    "#table(FinalS$toCheck <- (FP_1 | FP_2 | FP_3 | FP_4 | FP_5 | FP_6) & grepl(\"path\",FinalS$CLINSIG,ignore.case = TRUE)) #11 \n",
    "\n",
    "#View(FinalS[FinalS$toCheck,])\n",
    "sort(unique(FinalS$CLNDBN[FinalS$toCheck]))\n",
    "sort(unique(FinalS$Gene.refGene[FinalS$toCheck]))\n",
    "\n",
    "table(toReject1 <- FinalS$toCheck & !FinalS$n_alt_count%in%c(\"0\",\".\")) #1\n",
    "table(toReject2 <- FinalS$toCheck  & FinalS$PopFreq==\"reject\") #6\n",
    "table(toKeep <- FinalS$toCheck & !(toReject1 | toReject2)) #5\n",
    "table(toKeep <- toKeep | (FinalS$toCheck & FinalS$Gene.refGene%in%c(\"TP53\",\"IDH1\"))) #5\n",
    "\n",
    "#View(FinalS[toKeep,])\n",
    "\n",
    "FP <- (FP_1 | FP_2 | FP_3 | FP_4 | FP_5) & !toKeep\n",
    "#FP <- (FP_1 | FP_2 | FP_3 | FP_4 | FP_5 | FP_6) & !toKeep\n",
    "\n",
    "table(FP)\n",
    "# FALSE  TRUE \n",
    "#  1109  6459 \n",
    "\n",
    "FinalF <- FinalS[!FP,]\n",
    "FinalS <- data.frame(FP=\".\",FinalS,stringsAsFactors = FALSE)\n",
    "FinalS$FP[FP] <- \"reject\"\n",
    "table(FinalS$FP)\n",
    "#    . reject \n",
    "#  1109   6459 \n",
    "# FinalF <- FinalS[FinalS$FP==\".\",]\n",
    "nrow(FinalF) #1104\n",
    "rownames(FinalF) <- NULL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hotspot like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n",
      "[1] 8\n",
      "[1] 9\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 12\n",
      "[1] 13\n",
      "[1] 14\n",
      "[1] 15\n",
      "[1] 16\n",
      "[1] 17\n",
      "[1] 18\n",
      "[1] 19\n",
      "[1] 20\n",
      "[1] 21\n",
      "[1] 22\n",
      "[1] 23\n",
      "[1] 24\n",
      "[1] 25\n",
      "[1] 26\n",
      "[1] 27\n",
      "[1] 28\n",
      "[1] 29\n",
      "[1] 30\n",
      "[1] 31\n",
      "[1] 32\n",
      "[1] 33\n",
      "[1] 34\n",
      "[1] 35\n",
      "[1] 36\n",
      "[1] 37\n",
      "[1] 38\n",
      "[1] 39\n",
      "[1] 40\n",
      "[1] 41\n",
      "[1] 42\n",
      "[1] 43\n",
      "[1] 44\n",
      "[1] 45\n",
      "[1] 46\n",
      "[1] 47\n",
      "[1] 48\n",
      "[1] 49\n",
      "[1] 50\n",
      "[1] 51\n",
      "[1] 52\n",
      "[1] 53\n",
      "[1] 54\n",
      "[1] 55\n",
      "[1] 56\n",
      "[1] 57\n",
      "[1] 58\n",
      "[1] 59\n",
      "[1] 60\n",
      "[1] 61\n",
      "[1] 62\n",
      "[1] 63\n",
      "[1] 64\n",
      "[1] 65\n",
      "[1] 66\n",
      "[1] 67\n",
      "[1] 68\n",
      "[1] 69\n",
      "[1] 70\n",
      "[1] 71\n",
      "[1] 72\n",
      "[1] 73\n",
      "[1] 74\n",
      "[1] 75\n",
      "[1] 76\n",
      "[1] 77\n",
      "[1] 78\n",
      "[1] 79\n",
      "[1] 80\n",
      "[1] 81\n",
      "[1] 82\n",
      "[1] 83\n",
      "[1] 84\n",
      "[1] 85\n",
      "[1] 86\n",
      "[1] 87\n",
      "[1] 88\n",
      "[1] 89\n",
      "[1] 90\n",
      "[1] 91\n",
      "[1] 92\n",
      "[1] 93\n",
      "[1] 94\n",
      "[1] 95\n",
      "[1] 96\n",
      "[1] 97\n",
      "[1] 98\n",
      "[1] 99\n",
      "[1] 100\n",
      "[1] 101\n",
      "[1] 102\n",
      "[1] 103\n",
      "[1] 104\n",
      "[1] 105\n",
      "[1] 106\n",
      "[1] 107\n",
      "[1] 108\n",
      "[1] 109\n",
      "[1] 110\n",
      "[1] 111\n",
      "[1] 112\n",
      "[1] 113\n",
      "[1] 114\n",
      "[1] 115\n",
      "[1] 116\n",
      "[1] 117\n",
      "[1] 118\n",
      "[1] 119\n",
      "[1] 120\n",
      "[1] 121\n",
      "[1] 122\n",
      "[1] 123\n",
      "[1] 124\n",
      "[1] 125\n",
      "[1] 126\n",
      "[1] 127\n",
      "[1] 128\n",
      "[1] 129\n",
      "[1] 130\n",
      "[1] 131\n",
      "[1] 132\n",
      "[1] 133\n",
      "[1] 134\n",
      "[1] 135\n",
      "[1] 136\n",
      "[1] 137\n",
      "[1] 138\n",
      "[1] 139\n",
      "[1] 140\n",
      "[1] 141\n",
      "[1] 142\n",
      "[1] 143\n",
      "[1] 144\n",
      "[1] 145\n",
      "[1] 146\n",
      "[1] 147\n",
      "[1] 148\n",
      "[1] 149\n",
      "[1] 150\n",
      "[1] 151\n",
      "[1] 152\n",
      "[1] 153\n",
      "[1] 154\n",
      "[1] 155\n",
      "[1] 156\n",
      "[1] 157\n",
      "[1] 158\n",
      "[1] 159\n",
      "[1] 160\n",
      "[1] 161\n",
      "[1] 162\n",
      "[1] 163\n",
      "[1] 164\n",
      "[1] 165\n",
      "[1] 166\n",
      "[1] 167\n",
      "[1] 168\n",
      "[1] 169\n",
      "[1] 170\n",
      "[1] 171\n",
      "[1] 172\n",
      "[1] 173\n",
      "[1] 174\n",
      "[1] 175\n",
      "[1] 176\n",
      "[1] 177\n",
      "[1] 178\n",
      "[1] 179\n",
      "[1] 180\n",
      "[1] 181\n",
      "[1] 182\n",
      "[1] 183\n",
      "[1] 184\n",
      "[1] 185\n",
      "[1] 186\n",
      "[1] 187\n",
      "[1] 188\n",
      "[1] 189\n",
      "[1] 190\n",
      "[1] 191\n",
      "[1] 192\n",
      "[1] 193\n",
      "[1] 194\n",
      "[1] 195\n",
      "[1] 196\n",
      "[1] 197\n",
      "[1] 198\n",
      "[1] 199\n",
      "[1] 200\n",
      "[1] 201\n",
      "[1] 202\n",
      "[1] 203\n",
      "[1] 204\n",
      "[1] 205\n",
      "[1] 206\n",
      "[1] 207\n",
      "[1] 208\n",
      "[1] 209\n",
      "[1] 210\n",
      "[1] 211\n",
      "[1] 212\n",
      "[1] 213\n",
      "[1] 214\n",
      "[1] 215\n",
      "[1] 216\n",
      "[1] 217\n",
      "[1] 218\n",
      "[1] 219\n",
      "[1] 220\n",
      "[1] 221\n",
      "[1] 222\n",
      "[1] 223\n",
      "[1] 224\n",
      "[1] 225\n",
      "[1] 226\n",
      "\n",
      "   ABCA2    ABCB4   ABLIM2     ACPT     ADGB    AGXT2   ANGPT1    ARAP2 \n",
      "       2       20       20       20       18       15       18       18 \n",
      " ARHGDIA ARHGEF16    ATP9B   AVPR1A C16orf96 C1orf167  C1orf56 C20orf27 \n",
      "      20       20        2       20       15        2       20        2 \n",
      "  CDC25C   CHAMP1 COLGALT1     CUX1    CWH43   DNAJC6 EBNA1BP2     EGFR \n",
      "      17        3       20        4       20       19       20       40 \n",
      " EPB41L2    EPHB1       F8      GBA    GLP2R    GRIK4     GRM6   HAVCR1 \n",
      "      20       20       20        2       20       17       20        6 \n",
      "   HDAC1   HMGCS1    HOXD4    HTRA3   INO80D     IPO4     IPPK    KCNF1 \n",
      "      20        2        3       20       20       20        8       20 \n",
      "     KDR    KIF27    KIF3C  L3MBTL3    LRRC6    MAGIX   MALRD1    MAP1A \n",
      "      18       20       18        2       20       20       20        2 \n",
      "    MYO6    MYT1L   NANOS1  NEUROD4    NIM1K   NMNAT2     NPAT    NXPH4 \n",
      "      14       20        2       20       20        4       13       14 \n",
      "    OPN4   PCED1B    PDS5B   PHLDB1   PLXNA2   POU6F2   PRSS53     PSD2 \n",
      "       2       40        9       20        7        6        2        5 \n",
      "   PTPRF     PUM1   QRICH2  RANBP17    RBM47    RREB1     RTL9     RTP3 \n",
      "      20       20       20       16       20       19       20       20 \n",
      "  RUBCNL   SELPLG    SENP6    SHOX2  SHROOM2  SLC18A1   SLC2A1  SLC35A3 \n",
      "      20       20        2        6       15       20        2       11 \n",
      "   SOGA1    SPAG9    SRPRA      TEK   TOM1L2     TP53    TSHZ3   TUBA3C \n",
      "      20       20        6        4       20       20       20       20 \n",
      "  TXNRD2     TYW1    USP28     USP6   VPREB1      XPC    XRCC5  ZDHHC18 \n",
      "      20       20       20       20       20       20       12       20 \n",
      "   ZFHX2   ZNF131   ZNF185   ZNF480   ZNF726   ZNF772   ZNF831 \n",
      "       6        9       20       20       13       20       20 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "FinalF <- data.frame(Hotspot=\".\",FinalF,stringsAsFactors = FALSE)\n",
    "mut <- paste0(FinalF$Chr,\",\",FinalF$Start)\n",
    "Mut <- table(mut)\n",
    "Mut <- sort(Mut,decreasing=TRUE)\n",
    "Mut <- cbind(Mut,1)\n",
    "for(i in 1:nrow(Mut)){\n",
    "  print(i)\n",
    "  tmp <- unique(FinalF[mut==rownames(Mut)[i],\"Patient_ID\"])\n",
    "  # tmp <- unique(sapply(tmp,function(x) unlist(strsplit(x,\"_\"))[1]))\n",
    "  Mut[i,2] <- length(tmp)\n",
    "}\n",
    "\n",
    "for(i in 1:nrow(FinalF)){\n",
    "  FinalF[i,\"Hotspot\"] <- Mut[mut[i],2]\n",
    "}\n",
    "table(FinalF$Hotspot)\n",
    "\n",
    "# 1  10  12  13  14  15  16  17  18  19   2  20   3   4   5   6   7   8 \n",
    "# 42  10  12  13  28  45  48  51  18  38  22 700   9  12   5  12   7  32 \n",
    "\n",
    "#View(FinalF[FinalF$Hotspot!=1,])\n",
    "table(FinalF$Gene.refGene[FinalF$Hotspot!=1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# REDUNDANT MUTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"EGFR\"\n",
      "[1] \"PCED1B\"\n",
      "[1] \"ABCB4\"\n",
      "[1] \"ABLIM2\"\n",
      "[1] \"ACPT\"\n",
      "[1] \"ARHGDIA\"\n",
      "[1] \"ARHGEF16\"\n",
      "[1] \"AVPR1A\"\n",
      "[1] \"C1orf56\"\n",
      "[1] \"COLGALT1\"\n",
      "[1] \"CWH43\"\n",
      "[1] \"EBNA1BP2\"\n",
      "[1] \"EPB41L2\"\n",
      "[1] \"EPHB1\"\n",
      "[1] \"GLP2R\"\n",
      "[1] \"GRM6\"\n",
      "[1] \"HDAC1\"\n",
      "[1] \"HTRA3\"\n",
      "[1] \"INO80D\"\n",
      "[1] \"IPO4\"\n",
      "[1] \"KCNF1\"\n",
      "[1] \"KIF27\"\n",
      "[1] \"LRRC6\"\n",
      "[1] \"MAGIX\"\n",
      "[1] \"MALRD1\"\n",
      "[1] \"MYT1L\"\n",
      "[1] \"NEUROD4\"\n",
      "[1] \"NIM1K\"\n",
      "[1] \"PHLDB1\"\n",
      "[1] \"PTPRF\"\n",
      "[1] \"PUM1\"\n",
      "[1] \"QRICH2\"\n",
      "[1] \"RBM47\"\n",
      "[1] \"RTL9\"\n",
      "[1] \"RTP3\"\n",
      "[1] \"RUBCNL\"\n",
      "[1] \"SELPLG\"\n",
      "[1] \"SLC18A1\"\n",
      "[1] \"SOGA1\"\n",
      "[1] \"SPAG9\"\n",
      "[1] \"TOM1L2\"\n",
      "[1] \"TP53\"\n",
      "[1] \"TSHZ3\"\n",
      "[1] \"TUBA3C\"\n",
      "[1] \"TXNRD2\"\n",
      "[1] \"TYW1\"\n",
      "[1] \"USP28\"\n",
      "[1] \"USP6\"\n",
      "[1] \"VPREB1\"\n",
      "[1] \"XPC\"\n",
      "[1] \"ZDHHC18\"\n",
      "[1] \"ZNF185\"\n",
      "[1] \"ZNF480\"\n",
      "[1] \"ZNF772\"\n",
      "[1] \"ZNF831\"\n",
      "[1] \"DNAJC6\"\n",
      "[1] \"RREB1\"\n",
      "[1] \"ADGB\"\n",
      "[1] \"ANGPT1\"\n",
      "[1] \"ARAP2\"\n",
      "[1] \"KDR\"\n",
      "[1] \"KIF3C\"\n",
      "[1] \"CDC25C\"\n",
      "[1] \"GRIK4\"\n",
      "[1] \"ZNF131\"\n",
      "[1] \"AGXT2\"\n",
      "[1] \"C16orf96\"\n",
      "[1] \"SHROOM2\"\n",
      "[1] \"MYO6\"\n",
      "[1] \"NXPH4\"\n",
      "[1] \"ZNF726\"\n",
      "[1] \"SLC35A3\"\n",
      "[1] \"IPPK\"\n",
      "[1] \"NPAT\"\n",
      "[1] \"PDS5B\"\n",
      "[1] \"PLXNA2\"\n",
      "[1] \"HAVCR1\"\n",
      "[1] \"HMGCS1\"\n",
      "[1] \"POU6F2\"\n",
      "[1] \"SHOX2\"\n",
      "[1] \"SRPRA\"\n",
      "[1] \"ZFHX2\"\n",
      "[1] \"PSD2\"\n",
      "[1] \"CUX1\"\n",
      "[1] \"NMNAT2\"\n",
      "[1] \"TEK\"\n",
      "[1] \"CHAMP1\"\n",
      "[1] \"HOXD4\"\n",
      "[1] \"ABCA2\"\n",
      "[1] \"ATP9B\"\n",
      "[1] \"C1orf167\"\n",
      "[1] \"C20orf27\"\n",
      "[1] \"EMC7\"\n",
      "[1] \"HHIPL2\"\n",
      "[1] \"L3MBTL3\"\n",
      "[1] \"PRSS53\"\n",
      "[1] \"SENP6\"\n",
      "[1] \"SLC2A1\"\n",
      "[1] \"ADAMTS1\"\n",
      "[1] \"ALDH2\"\n",
      "[1] \"AMOT\"\n",
      "[1] \"APC2\"\n",
      "[1] \"ARHGEF40\"\n",
      "[1] \"BHLHE22\"\n",
      "[1] \"BTN3A2\"\n",
      "[1] \"C11orf80\"\n",
      "[1] \"C17orf78\"\n",
      "[1] \"CACNA1A\"\n",
      "[1] \"CACNA1G\"\n",
      "[1] \"CAPN14\"\n",
      "[1] \"CAPN3\"\n",
      "[1] \"CASC3\"\n",
      "[1] \"CDADC1\"\n",
      "[1] \"CENPB\"\n",
      "[1] \"CGREF1\"\n",
      "[1] \"CHRNA3\"\n",
      "[1] \"CNNM1\"\n",
      "[1] \"CPQ\"\n",
      "[1] \"CYP2W1\"\n",
      "[1] \"DCBLD2\"\n",
      "[1] \"DGCR2\"\n",
      "[1] \"DNAJC16\"\n",
      "[1] \"EIF2B4\"\n",
      "[1] \"ESYT2\"\n",
      "[1] \"FBXO17\"\n",
      "[1] \"FCER2\"\n",
      "[1] \"FLNA\"\n",
      "[1] \"FOS\"\n",
      "[1] \"FSCN3\"\n",
      "[1] \"GJA9\"\n",
      "[1] \"HSD17B10\"\n",
      "[1] \"IARS\"\n",
      "[1] \"IL32\"\n",
      "[1] \"ING2\"\n",
      "[1] \"INVS\"\n",
      "[1] \"JAKMIP3\"\n",
      "[1] \"JARID2\"\n",
      "[1] \"KLHL29\"\n",
      "[1] \"LHFPL1\"\n",
      "[1] \"LOXL1\"\n",
      "[1] \"LRRC37B\"\n",
      "[1] \"MAGI1\"\n",
      "[1] \"MALT1\"\n",
      "[1] \"MAP3K6\"\n",
      "[1] \"MGAM2\"\n",
      "[1] \"MROH1\"\n",
      "[1] \"MTMR3\"\n",
      "[1] \"MTUS2\"\n",
      "[1] \"MYLK4\"\n",
      "[1] \"MYO5C\"\n",
      "[1] \"NAPA\"\n",
      "[1] \"NBEA\"\n",
      "[1] \"NEFM\"\n",
      "[1] \"NHEJ1\"\n",
      "[1] \"NOTCH3\"\n",
      "[1] \"NPBWR1\"\n",
      "[1] \"NTRK3\"\n",
      "[1] \"PALLD\"\n",
      "[1] \"PEX11B\"\n",
      "[1] \"PFKFB2\"\n",
      "[1] \"PGS1\"\n",
      "[1] \"PLCB1\"\n",
      "[1] \"PLEKHG5\"\n",
      "[1] \"PSMD8\"\n",
      "[1] \"RARA\"\n",
      "[1] \"RARRES2\"\n",
      "[1] \"RASA1\"\n",
      "[1] \"RBP3\"\n",
      "[1] \"RGPD2\"\n",
      "[1] \"SCUBE1\"\n",
      "[1] \"SEC31B\"\n",
      "[1] \"SEMA3C\"\n",
      "[1] \"SETDB1\"\n",
      "[1] \"SGSM3\"\n",
      "[1] \"SH3RF1\"\n",
      "[1] \"SHB\"\n",
      "[1] \"SLC2A3\"\n",
      "[1] \"SLC7A8\"\n",
      "[1] \"SMR3B\"\n",
      "[1] \"SOX1\"\n",
      "[1] \"SPG7\"\n",
      "[1] \"TBC1D1\"\n",
      "[1] \"TBX3\"\n",
      "[1] \"TDRD10\"\n",
      "[1] \"TENM2\"\n",
      "[1] \"THOC3\"\n",
      "[1] \"USP20\"\n",
      "[1] \"USP22\"\n",
      "[1] \"WDR63\"\n",
      "[1] \"WNT9A\"\n",
      "[1] \"XRN1\"\n",
      "[1] \"ZNF212\"\n",
      "[1] \"ZNF500\"\n",
      "[1] \"ZNF551\"\n",
      "[1] \"ZXDC\"\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "rownames(FinalF) <- NULL\n",
    "id <- paste0(FinalF$Tumor_ID,\"|\",FinalF$Gene.refGene)\n",
    "table(red <- duplicated(id))\n",
    "# FALSE  TRUE \n",
    "#  112   992 \n",
    "red <- id%in%id[red]\n",
    "FinalF <- data.frame(red,FinalF,stringsAsFactors = FALSE)\n",
    "#View(FinalF[FinalF$red==TRUE,c(\"Patient_ID\",\"caller\",\"callerN\",\"Chr\",\"Start\",\"End\",\"Ref\",\"Alt\",\"Gene.refGene\")])\n",
    "toRemove <- FinalF$red & FinalF$callerN==1\n",
    "#View(FinalF[toRemove,c(\"Patient_ID\",\"caller\",\"callerN\",\"Chr\",\"Start\",\"End\",\"Ref\",\"Alt\",\"Gene.refGene\")])\n",
    "FinalF <- FinalF[!toRemove,]\n",
    "rownames(FinalF) <- NULL\n",
    "nrow(FinalF) #1078\n",
    "\n",
    "# Mutation recurrence (correct to count only one mutation for patient)\n",
    "genes <- sapply(FinalF[,\"Gene.refGene\"],function(x) unlist(strsplit(x,\",\"))[1])\n",
    "MutGenes <- table(genes)\n",
    "MutGenes <- sort(MutGenes,decreasing=TRUE)\n",
    "MutGenes <- cbind(MutGenes,1)\n",
    "for(i in 1:nrow(MutGenes)){\n",
    "  print(rownames(MutGenes)[i])\n",
    "  tmp <- FinalF[genes==rownames(MutGenes)[i],\"Patient_ID\"]\n",
    "  tmp <- unique(tmp)\n",
    "  MutGenes[i,2] <- length(tmp)\n",
    "}\n",
    "FinalF <- data.frame(FinalF,GeneRec=1,GenePzRec=1,stringsAsFactors = FALSE)\n",
    "for(i in 1:nrow(FinalF)){\n",
    "  FinalF[i,\"GeneRec\"] <- MutGenes[genes[i],1]\n",
    "  FinalF[i,\"GenePzRec\"] <- MutGenes[genes[i],2]\n",
    "}\n",
    "table(FinalF$GeneRec)\n",
    "table(FinalF$GenePzRec)\n",
    "#View(FinalF[FinalF$GenePzRec!=1,])\n",
    "\n",
    "mutCounts <- sort(table(FinalF$Tumor_ID))\n",
    "\n",
    "tmp <- colnames(FinalF)\n",
    "sel <- c(\"Patient_ID\", #\"Tumor_ID\", \"Normal_ID\",\n",
    "         \"caller\",\t\"callerN\",\t\"Type\",\t\"Chr\",\t\"Start\",\t\"End\", \"Ref\",\t\"Alt\",\t\n",
    "         \"Gene.refGene\",\t\"Description\",\t\"Func.refGene\",\t\"GeneDetail.refGene\",\"snpEff_func\",\t\"HGVSc\",\t\"HGVSp\",\t\"Interpro_domain\",\t\n",
    "         \"Pathogenicity\",\t\"MissenseDamaging\",\t\"SIFT_pred\",\t\"Polyphen2_HDIV_pred\",\t\"MutationTaster_pred\",\t\"PROVEAN_pred\",\t\n",
    "         \"census\",\t\"censusGermline\",\t\"nature\",\t\"natureGermline\",\t\"oncoKB\",\t\"oncoKBmut\",\t\"smgOcc\",\t\n",
    "         \"FPfilter\",\t\"FPfail\",\t\"MaskedFilter\", \"PON_filter\", \"PON_filter_pass\", \"Hotspot\",\n",
    "        # \"VN_occurrences\",\"vn\",\n",
    "         \"avsnp150\",\t\"ExAC_ALL\",\t\"X1000g2015aug_all\", \"esp6500siv2_all\",\t\"Kaviar_AF\",\t\"HRC_AF\",\t\n",
    "         \"cosmic80\",\t\"CLINSIG\",\t\"CLNDBN\",\t\"CLNACC\",\t\"CLNDSDB\",\t\"CLNDSDBID\",\n",
    "         \"t_depth\",\t\"t_ref_count\",\t\"t_alt_count\", \"n_depth\",\t\"n_ref_count\",\t\"n_alt_count\",\t\"t_vaf\",\t\"n_vaf\",\n",
    "         \"cBioId\",\t\"cBioAAchange\",\t\"cBioKnownVar\",\t\"cBioStudyId\",\t\"cBioSamplesNum\",\t\"cBioFreq\",\n",
    "         \"GeneRec\",\t\"GenePzRec\")\n",
    "setdiff(sel, tmp)\n",
    "setdiff(tmp, sel)  \n",
    "FinalF <- FinalF[,sel]\n",
    "\n",
    "save(Final,FinalS,FinalF,file=\"Final_no_PoN_filter.rda\")\n",
    "FinalS <- FinalS[,!colnames(FinalS)%in%c(\"snpEff\")]\n",
    "write.table(FinalS,file = \"ABT414_SomaticVariants.txt\",sep=\"\\t\",row.names = FALSE)\n",
    "write.table(FinalF,file = \"ABT414_SomaticVariants_FPfilt.txt\",sep=\"\\t\",row.names = FALSE)\n",
    "write.table(FinalF$Gene.refGene,file = \"ABT414_SomaticVariants_FPfilt_genes.txt\",sep=\"\\t\",row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
